# Coral-associated bacteria demonstrate phylosymbiosis and cophylogeny

{cite:t}`Pollock_2018`

## Notes pulled from paper

> "Modern corals harbor complex communities of microorganisms, including dinoflagellates, fungi, bacteria, and archaea which are collectively termed the coral microbiome"

> "Coral mucus, tissue, and skeleton show distinct microbial community composition"

> "comparisons of six species of coral and an octocoral outgroup found microbiome similarities that seemed to better align with morphology than phylogeny"

> "sequenced 691 16S rRNA gene libraries from these samples = ["DNA samples from the mucus, tissue, and skeleton of phylogenetically diverse scleractinian species, as well as selected outgroups and environmental references"] [...] paired these microbiome data with a multigene molecular phylogeny of scleractinian corals, coral functional traits from the Coral Trait Database, and extensive in-situ metadata (Methods)"

> "Coral tissue, mucus, and skeleton microbiomes differed in richness (Fig. 1a) and microbiome composition (Fig. 1b). Surprisingly, the coral endolithic skeleton was richer in microbial diversity than the tissue microbiome"

> **"Overall, mucus microbiomes were notable for their relative stability between colonies and their high abundance of core vs. variable microbes"**

> "For each environmental and host parameter, we tested its relative influence on coral mucus, tissue, and skeletal microbiomes (Fig. 1c). We then clustered host and environmental parameters in terms of their effects on the microbiome across compartments. Intrinsic host-based traits clustered separately from environmental traits."

> **"environmental factors (e.g., season, temperature and turf algal competition) had a stronger influence on mucus microbiomes than tissue or skeleton microbiomes"**

> "while host species influences the microbiome across anatomy, the extent of host vs. environmental influence on coral microbiomes is not consistent in coral mucus, tissue, and skeleton"

> "the **latitude of the sampling location significantly influenced the richness and composition of coral microbiomes**. Moving away from the equator, coral microbiomes became less rich. This effect was **significant for coral mucus and tissue**, but not skeleton microbiomes, even after accounting for the uneven distribution of species across locations"

> "Effects of coral relative size on microbiome richness were significant in coral tissue and skeleton, but not mucus, after accounting for phylogeny"

> "Using the phylogenetically-aware Weighted UniFrac method deemphasized fine variation at the tips of the microbial tree, resulting in a **significant signal of phylosymbiosis in skeleton, but not tissue or mucus microbiomes**"

> "Mucus microbiomes, on the other hand, did not become more dissimilar as host phylogenetic distance increased past the second distance class"

> "like community composition, richness was significantly more similar among closely related corals than expected by chance"

> "the scales of phylogenetic divergence at which these effects appear, suggest that **the radiation of modern reef-building coral families (between roughly 25 and 65 mya) was accompanied by large changes in microbiome richness**, with changes continuing to accumulate during more recent speciation events"

> "the phylogenetic histories of corals partially constrain the composition of their tissue and skeletal microbiomes and the richness of all coral compartments"

> "most microbial genera showed host-specific abundance patterns"

> "most coral-associated microbes were correlated with host identity, not host phylogeny"

> "**Mucus microbiomes showed fewer genera (1.6%) associated with host phylogeny than tissue, while skeletal microbiomes showed more (4.9%**)"

> "Phylogenetic GLMMs estimated that the abundances of between 62% (tissue) and 75% (mucus) of microbial genera were significantly correlated with at least one host mitotype (host identity), depending on the compartment analyzed"

> "All but two of the families tested had some form of host specificity, with members either associated with particular coral mitotypes (representing species or genera) or particular regions of the coral tree"

> - "host identity by bacterial identity;" `<-` [...] *'were only significant in three bacterial groups: Clostridiaceae (mucus), unclassified Myxococcales (mucus), and unclassified Kiloniellales (mucus and tissue)'*
> - "host identity by bacterial phylogeny;" `<-` [...] *'individual coral mitotypes were significantly associated with clades of related bacteria'*
> - "host phylogeny by bacterial identity; or"
> - "host phylogeny by bacterial phylogeny (i.e., cophylogeny)" `<-` [...] *'Four bacterial groups exhibited significant cophylogenetic effects [...] Clostridiaceae, Endozoicomonas-like bacteria (Endozoicomonaceae in Greengenes), unclassified Kiloniellales, and unclassified Myxococcales'*

> "the majority of microbial genera show significant host-specificity"

> "**host specificity in the mucus microbiome was limited to relatively recent divergences and was not significantly structured by larger scales of host phylogeny**"

> "In contrast to the patterns in the mucus, the coral skeleton, which has been less intensively studied than mucus and tissue, showed both the greatest microbiome richness and the strongest signal of long-term phylosymbiosis"

> "although many coral-associated bacteria are host-specific, and the overall composition of coral microbiomes tracks phylogeny (i.e., phylosymbiosis), only a select minority of coral-associated bacterial families show cophylogenetic signals consistent with long-term host-microbe codiversification"

> "The microbial families that show signs of cophylogeny may be associated with important host functions that have led to stable symbiotic relationships across extended evolutionary time"
