

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Pilot 3 &#8212; Coral Bleaching - BIOL 417</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=ac02cc09edc035673794" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=ac02cc09edc035673794" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=ac02cc09edc035673794" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=ac02cc09edc035673794" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/drawio.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=ac02cc09edc035673794" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=ac02cc09edc035673794" />
  <script src="../../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=ac02cc09edc035673794"></script>

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'src/3_pilot_3/download';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Tree finding" href="tree_finding.html" />
    <link rel="prev" title="Pilot 1 &amp; 2" href="../3_pilot_1_2/index.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../readme.html">
  
  
  
  
  
    <p class="title logo__title">Coral Bleaching - BIOL 417</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../readme.html">
                    Welcome to BIOL 417 101 Capstone!
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Exploring Literature</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../1_exploratory/index.html">Proposal ideas</a></li>
<li class="toctree-l1"><a class="reference internal" href="../1_exploratory/datasets.html">List of used datasets for analysis</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Proposal</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../2_proposal/index.html">Proposal</a></li>
<li class="toctree-l1"><a class="reference internal" href="../2_proposal/literature.html">Literature Review</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Pilot analysis</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../3_pilot_1_2/index.html">Pilot 1 &amp; 2</a></li>
<li class="toctree-l1 current active has-children"><a class="current reference internal" href="#">Pilot 3</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="tree_finding.html">Tree finding</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Full analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../4_full_analysis/download.html">Final analysis download</a></li>
<li class="toctree-l1"><a class="reference internal" href="../4_full_analysis/download_genome.html">Download genomes</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">End</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../references.html">References</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/Mapleia/BIOL417_Coral" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/src/3_pilot_3/download.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Pilot 3</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#setup-the-environment">Setup the environment</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#sample-search-to-download-from-ncbi">Sample search to download from NCBI</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#download-sequences-for-our-organism-list">Download sequences for our organism list</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#align-the-sequences">Align the sequences</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="pilot-3">
<h1>Pilot 3<a class="headerlink" href="#pilot-3" title="Permalink to this heading">#</a></h1>
<p>I wanted to see if I can use more than the 16S rRNA gene to identify the species of the bacteria. I first searched online and found that there was a further breakdown in taxa. Taking note of the list of organisms, I scripted a search for Endozoicomonas &amp; coral. This included a lot of whole genome shotgunning, which in the end turned out to be useless for further analysis, and took a long time to align and tree search.</p>
<p>We are using the package biopython to download nucleotide datasets from NCBI’s Genbank, then prep those to align later. The guide on downloading from NCBI is from <a class="reference external" href="https://notebook.community/widdowquinn/Notebooks-Bioinformatics/Biopython_NCBI_Entrez_downloads">this guide</a>.</p>
<section id="setup-the-environment">
<h2>Setup the environment<a class="headerlink" href="#setup-the-environment" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span><span class="p">,</span> <span class="n">SeqIO</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="c1"># from Bio.Align.Applications import ClustalwCommandline</span>
<span class="kn">import</span> <span class="nn">os</span>
</pre></div>
</div>
</div>
</div>
<p>Setup the <code class="docutils literal notranslate"><span class="pre">ncbi_download</span></code> directory, and check that the list of organisms is in the file <a class="reference download internal" download="" href="../../_downloads/3ebbed5bb6034d7882fdf5bfd03d72e7/organism_list.txt"><span class="xref download myst"><code class="docutils literal notranslate"><span class="pre">organisms.txt</span></code></span></a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a new directory (if needed) for output/downloads</span>
<span class="n">path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">realpath</span><span class="p">(</span><span class="s2">&quot;..</span><span class="se">\\</span><span class="s2">..</span><span class="se">\\</span><span class="s2">datasets</span><span class="se">\\</span><span class="s2">shotgun_sequences</span><span class="se">\\</span><span class="s2">ncbi_downloads&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
<span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">out_amount</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>

<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">out_amount</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;No organism list in ncbi_downloads.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>E:\Kaede\Documents\GitHub\BIOL417_CORAL\datasets\shotgun_sequences\ncbi_downloads
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;kaedeito@student.ubc.ca&quot;</span>
<span class="c1"># This line sets the name of the tool that is making the queries</span>
<span class="n">Entrez</span><span class="o">.</span><span class="n">tool</span> <span class="o">=</span> <span class="s2">&quot;alignment.ipynb&quot;</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="sample-search-to-download-from-ncbi">
<h2>Sample search to download from NCBI<a class="headerlink" href="#sample-search-to-download-from-ncbi" title="Permalink to this heading">#</a></h2>
<p>Here is an example using the generic search term of “Endozoicomonas” and “coral” in NCBI’s Genbank nucleotide database.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">Entrez.esearch(...)</span></code> function is documented <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ESearch">here</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio.Entrez.Parser</span> <span class="kn">import</span> <span class="n">DictionaryElement</span><span class="p">,</span> <span class="n">IntegerElement</span><span class="p">,</span> <span class="n">ListElement</span><span class="p">,</span> <span class="n">NoneElement</span>


<span class="k">def</span> <span class="nf">search_sample_coral_bacteria</span><span class="p">():</span>
  <span class="n">term</span> <span class="o">=</span> <span class="s2">&quot;Endozoicomonas[Organism] AND coral[All Fields] AND biomol_genomic[PROP]&quot;</span>
  <span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span>
    <span class="n">term</span><span class="o">=</span><span class="n">term</span><span class="p">,</span>
    <span class="n">retmax</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">idtype</span><span class="o">=</span><span class="s2">&quot;acc&quot;</span><span class="p">,</span>
  <span class="p">)</span>

  <span class="c1"># This line converts the returned information from NCBI into a form we</span>
  <span class="c1"># can use, as before.</span>
  <span class="n">record</span><span class="p">:</span> <span class="n">DictionaryElement</span> <span class="o">|</span> <span class="n">ListElement</span> <span class="o">|</span> <span class="n">IntegerElement</span> <span class="o">|</span> <span class="n">NoneElement</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
  <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">DictionaryElement</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="p">[</span><span class="s1">&#39;IdList&#39;</span><span class="p">])</span>

<span class="n">search_sample_coral_bacteria</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;NZ_CP120717.1&#39;, &#39;CP120717.1&#39;, &#39;NZ_JAHHPM000000000.1&#39;, &#39;NZ_JAHHPM010000603.1&#39;, &#39;NZ_JAHHPM010000602.1&#39;, &#39;NZ_JAHHPM010000601.1&#39;, &#39;NZ_JAHHPM010000600.1&#39;, &#39;NZ_JAHHPM010000599.1&#39;, &#39;NZ_JAHHPM010000598.1&#39;, &#39;NZ_JAHHPM010000597.1&#39;]
</pre></div>
</div>
</div>
</div>
</section>
<section id="download-sequences-for-our-organism-list">
<h2>Download sequences for our organism list<a class="headerlink" href="#download-sequences-for-our-organism-list" title="Permalink to this heading">#</a></h2>
<p>From a file called “organism_list.txt”, read the list of organisms to search.</p>
<p>Add the search term “coral” to the end of each organism name, and limit to 10 results per taxon. Also limit the search to specific range of nucleotide length.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">list_orgs</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;../../datasets/shotgun_sequences/organism_list.txt&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span>

<span class="k">def</span> <span class="nf">load_records</span><span class="p">():</span>
  <span class="n">list_of_accession_id</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

  <span class="n">min_seq_length</span> <span class="o">=</span> <span class="mi">1200</span>
  <span class="n">max_seq_length</span> <span class="o">=</span> <span class="mi">10000</span>

  <span class="c1"># limit length of sequence</span>
  <span class="n">length_search</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;(</span><span class="si">{</span><span class="n">min_seq_length</span><span class="si">}</span><span class="s2">[SLEN] : </span><span class="si">{</span><span class="n">max_seq_length</span><span class="si">}</span><span class="s2">[SLEN])&quot;</span>

  <span class="k">for</span> <span class="n">org</span> <span class="ow">in</span> <span class="n">list_orgs</span><span class="p">:</span>
    <span class="n">term</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;(</span><span class="se">\&quot;</span><span class="si">{</span><span class="n">org</span><span class="si">}</span><span class="se">\&quot;</span><span class="s2">[Organism] OR </span><span class="si">{</span><span class="n">org</span><span class="si">}</span><span class="s2">[All Fields]) AND coral[All Fields] AND </span><span class="si">{</span><span class="n">length_search</span><span class="si">}</span><span class="s2"> AND biomol_genomic[PROP]&quot;</span>
    <span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span>
      <span class="n">term</span><span class="o">=</span><span class="n">term</span><span class="p">,</span>
      <span class="n">retmax</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
      <span class="n">idtype</span><span class="o">=</span><span class="s2">&quot;acc&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">DictionaryElement</span><span class="p">):</span>
      <span class="n">list_of_accession_id</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">record</span><span class="p">[</span><span class="s1">&#39;IdList&#39;</span><span class="p">])</span>

  <span class="k">return</span> <span class="n">list_of_accession_id</span>
</pre></div>
</div>
</div>
</div>
<p>A class to hold the search results, and a function to print the results to a fasta file.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">dataclasses</span> <span class="kn">import</span> <span class="n">dataclass</span>
<span class="kn">from</span> <span class="nn">io</span> <span class="kn">import</span> <span class="n">TextIOWrapper</span>

<span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">OrganismAccession</span><span class="p">:</span>
  <span class="n">ids</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span>
  <span class="n">organism</span><span class="p">:</span> <span class="nb">str</span>
  <span class="n">handle</span><span class="p">:</span> <span class="n">TextIOWrapper</span>
  <span class="n">folder</span><span class="p">:</span> <span class="nb">str</span>
  <span class="n">record</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">]</span>

  <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ids</span><span class="p">,</span> <span class="n">organism</span><span class="p">):</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">ids</span> <span class="o">=</span> <span class="n">ids</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">organism</span> <span class="o">=</span> <span class="n">organism</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">folder</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;..</span><span class="se">\\</span><span class="s2">..</span><span class="se">\\</span><span class="s2">datasets</span><span class="se">\\</span><span class="s2">shotgun_sequences</span><span class="se">\\</span><span class="s2">ncbi_downloads</span><span class="se">\\</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">organism</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">folder</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">ids</span><span class="p">,</span> <span class="n">rettype</span><span class="o">=</span><span class="s2">&quot;fasta&quot;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s2">&quot;text&quot;</span><span class="p">)</span>

  <span class="k">def</span> <span class="nf">download_save</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">folder</span><span class="si">}</span><span class="se">\\</span><span class="s2">nucleotide.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_handle</span><span class="p">:</span>
      <span class="n">out_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">handle</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
      <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">organism</span><span class="si">}</span><span class="s2"> Saved&quot;</span><span class="p">)</span>

  <span class="k">def</span> <span class="nf">get_record</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">folder</span><span class="si">}</span><span class="se">\\</span><span class="s2">nucleotide.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">download_sequences</span><span class="p">():</span>
<span class="w">  </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">  Download sequences from NCBI from the handle set by load_records()</span>
<span class="sd">  and save them in a folder named after the organism.</span>
<span class="sd">  Load the SeqRecords from the saved fasta file.</span>
<span class="sd">  &quot;&quot;&quot;</span>
  <span class="n">organisms_records</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

  <span class="n">len_org</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">list_orgs</span><span class="p">)</span>
  <span class="n">list_of_accession_id</span> <span class="o">=</span> <span class="n">load_records</span><span class="p">()</span>
  <span class="n">len_results</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">list_of_accession_id</span><span class="p">)</span>
  <span class="k">if</span> <span class="n">len_org</span> <span class="o">!=</span> <span class="n">len_results</span><span class="p">:</span>
    <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;Length of organism list and results list are not the same&quot;</span><span class="p">)</span>

  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">len_org</span><span class="p">):</span>
    <span class="n">org</span> <span class="o">=</span> <span class="n">list_orgs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
    <span class="n">ids</span> <span class="o">=</span> <span class="n">list_of_accession_id</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
    <span class="n">organism_accession</span> <span class="o">=</span> <span class="n">OrganismAccession</span><span class="p">(</span><span class="n">ids</span><span class="p">,</span> <span class="n">org</span><span class="p">)</span>
    <span class="n">organism_accession</span><span class="o">.</span><span class="n">download_save</span><span class="p">()</span>
    <span class="n">organism_accession</span><span class="o">.</span><span class="n">get_record</span><span class="p">()</span>
    <span class="n">organisms_records</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">organism_accession</span><span class="o">.</span><span class="n">records</span><span class="p">)</span>

  <span class="k">return</span> <span class="n">organisms_records</span>


<span class="k">def</span> <span class="nf">remove_duplicates</span><span class="p">(</span><span class="n">organisms_records</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">]):</span>
<span class="w">  </span><span class="sd">&quot;&quot;&quot;&quot;</span>
<span class="sd">  Remove duplicate records from the list of SeqRecords using ID.</span>
<span class="sd">  &quot;&quot;&quot;</span>
  <span class="n">accession_ids</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
  <span class="n">duplicates</span> <span class="o">=</span> <span class="p">[]</span>
  <span class="n">duplicate_ids</span> <span class="o">=</span> <span class="p">[]</span>

  <span class="n">retained_records</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
  <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">fasta</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">organisms_records</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">fasta</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">accession_ids</span><span class="p">:</span>
      <span class="n">duplicates</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
      <span class="n">duplicate_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="n">retained_records</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fasta</span><span class="p">)</span>
    <span class="n">accession_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>

  <span class="k">return</span> <span class="n">retained_records</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">organisms_records</span> <span class="o">=</span> <span class="n">download_sequences</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Endozoicomonas sp. ONNA2 Saved
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Endozoicomonas acroporae Saved
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Endozoicomonas sp. ONNA1 Saved
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Endozoicomonas sp. SESOKO3 Saved
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Endozoicomonas sp. SESOKO2 Saved
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Endozoicomonas sp. SESOKO4 Saved
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Endozoicomonas sp. YOMI1 Saved
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Endozoicomonas sp. SESOKO1 Saved
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Endozoicomonas sp. ISHI1 Saved
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Endozoicomonas montiporae Saved
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Endozoicomonas sp. G2_1 Saved
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Endozoicomonas sp. Saved
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Endozoicomonas sp. G2_2 Saved
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Endozoicomonas coralli Saved
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Endozoicomonas gorgoniicola Saved
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Endozoicomonas euniceicola Saved
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">organisms_records</span> <span class="o">=</span> <span class="n">remove_duplicates</span><span class="p">(</span><span class="n">organisms_records</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">organisms_records</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s2">&quot;../../outputs/pilot_3_shotgun&quot;</span><span class="p">,</span> <span class="s2">&quot;all_unaligned.fasta&quot;</span><span class="p">),</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Done writing unaligned fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Done writing unaligned fasta
</pre></div>
</div>
</div>
</div>
</section>
<section id="align-the-sequences">
<h2>Align the sequences<a class="headerlink" href="#align-the-sequences" title="Permalink to this heading">#</a></h2>
<p>I had MUSCLE v5.1 <span id="id1">[<a class="reference internal" href="../references.html#id30" title="Robert C. Edgar. High-accuracy alignment ensembles enable unbiased assessments of sequence homology and phylogeny. bioRxiv, 2022. URL: https://www.biorxiv.org/content/early/2022/08/01/2021.06.20.449169, arXiv:https://www.biorxiv.org/content/early/2022/08/01/2021.06.20.449169.full.pdf, doi:10.1101/2021.06.20.449169.">Edgar, 2022</a>]</span> align <a class="reference download internal" download="" href="../../_downloads/7434d4febbc90a4f1b0bb673218ec18b/all_unaligned.fasta"><span class="xref download myst">them</span></a>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>muscle5.1<span class="w"> </span>-input<span class="w"> </span>../../outputs/pilot_3_shotgun/all_unaligned.fasta<span class="w"> </span>-output<span class="w"> </span>../../outputs/pilot_3_shotgun/local/muscle/all_aligned.fasta
</pre></div>
</div>
</section>
<div class="toctree-wrapper compound">
</div>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./src\3_pilot_3"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  <!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="../3_pilot_1_2/index.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Pilot 1 &amp; 2</p>
      </div>
    </a>
    <a class="right-next"
       href="tree_finding.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Tree finding</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#setup-the-environment">Setup the environment</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#sample-search-to-download-from-ncbi">Sample search to download from NCBI</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#download-sequences-for-our-organism-list">Download sequences for our organism list</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#align-the-sequences">Align the sequences</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Kaede Ito
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=ac02cc09edc035673794"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=ac02cc09edc035673794"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>