

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Pilot 1 &amp; 2 &#8212; Coral Bleaching - BIOL 417</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=ac02cc09edc035673794" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=ac02cc09edc035673794" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=ac02cc09edc035673794" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=ac02cc09edc035673794" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/drawio.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=ac02cc09edc035673794" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=ac02cc09edc035673794" />
  <script src="../../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=ac02cc09edc035673794"></script>

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'src/3_pilot_1_2/index';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Pilot 3" href="../3_pilot_3/download.html" />
    <link rel="prev" title="Literature Review" href="../2_proposal/literature.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../readme.html">
  
  
  
  
  
    <p class="title logo__title">Coral Bleaching - BIOL 417</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../readme.html">
                    Welcome to BIOL 417 101 Capstone!
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Exploring Literature</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../1_exploratory/index.html">Proposal ideas</a></li>
<li class="toctree-l1"><a class="reference internal" href="../1_exploratory/datasets.html">List of used datasets for analysis</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Proposal</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../2_proposal/index.html">Proposal</a></li>
<li class="toctree-l1"><a class="reference internal" href="../2_proposal/literature.html">Literature Review</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Pilot analysis</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Pilot 1 &amp; 2</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../3_pilot_3/download.html">Pilot 3</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../3_pilot_3/tree_finding.html">Tree finding</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Full analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../4_full_analysis/download.html">Final analysis download</a></li>
<li class="toctree-l1"><a class="reference internal" href="../4_full_analysis/download_genome.html">Download genomes</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">End</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../references.html">References</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/Mapleia/BIOL417_Coral" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/src/3_pilot_1_2/index.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Pilot 1 & 2</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#basic-analysis">Basic analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#wrangle-data">Wrangle data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pike-et-al-2013-basic-analysis">Pike et al (2013) basic analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pilot-analysis-1-2">Pilot Analysis (1 &amp; 2)</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pilot-1">Pilot #1</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pilot-2">Pilot #2</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="pilot-1-2">
<h1>Pilot 1 &amp; 2<a class="headerlink" href="#pilot-1-2" title="Permalink to this heading">#</a></h1>
<p>I’ve pivoted away from the initial idea of comparing with plasmids, and instead I’m going to focus on the 16S rRNA gene.</p>
<section id="basic-analysis">
<h2>Basic analysis<a class="headerlink" href="#basic-analysis" title="Permalink to this heading">#</a></h2>
<p>Based on the paper <span id="id1">Pike <em>et al.</em> [<a class="reference internal" href="../references.html#id28" title="Rebecca E. Pike, Brad Haltli, and Russell G. Kerr. Description of endozoicomonas euniceicola sp. nov. and endozoicomonas gorgoniicola sp. nov., bacteria isolated from the octocorals eunicea fusca and plexaura sp., and an emended description of the genus endozoicomonas. International Journal of Systematic and Evolutionary Microbiology, 63(Pt_11):4294–4302, nov 2013. doi:10.1099/ijs.0.051490-0.">2013</a>]</span>, I wanted to expand on their phylogenetic analysis that mainly used ME (minimum evolution) method. I wanted to see if I could replicate their results using a different method, and also see if I could improve on their results by adding more 16S sequences.</p>
<ol class="arabic">
<li><p>They took their 16S ribosomal RNA samples and sequenced them.</p></li>
<li><p>They used BLASTN to quantify phylogenetic relatedness of sequences within GenBank.</p></li>
<li><p>They also used EzTaxon (now called EzBioCloud) to identify the closest relatives of their sequences.</p></li>
<li><p>They aligned using BioEdit v7.0.5.3.</p>
<blockquote>
<div><p>BioEdit is (was) a sequence alignment editor written for Windows 95/98/NT/2000/XP/7. It is free for academic and non-profit use. It is no longer supported, but the last version is still available for download. It is not available for Mac or Linux. <span id="id2">[<a class="reference internal" href="../references.html#id31" title="Tom A Hall and others. Bioedit: a user-friendly biological sequence alignment editor and analysis program for windows 95/98/nt. In Nucleic acids symposium series, volume 41, 95–98. Oxford, 1999.">Hall and others, 1999</a>]</span></p>
</div></blockquote>
</li>
<li><p>They used MEGA v4.0 (2007) <span id="id3">[<a class="reference internal" href="../references.html#id27" title="K. Tamura, J. Dudley, M. Nei, and S. Kumar. MEGA4: molecular evolutionary genetics analysis (MEGA) software version 4.0. Molecular Biology and Evolution, 24(8):1596–1599, apr 2007. doi:10.1093/molbev/msm092.">Tamura <em>et al.</em>, 2007</a>]</span> to construct a phylogenetic tree using the:</p>
<p>Generally:</p>
<ul class="simple">
<li><p>Neighbor-joining method: for initial tree construction</p></li>
<li><p>bootstrapping with 1000 replications</p></li>
<li><p>distance calculated with Jukes-Cantor model</p></li>
</ul>
<p>Method 1:</p>
<ul class="simple">
<li><p>minimum evolution method: close-neighbour-interchange algorithm &#64; level 1</p></li>
<li><p>complete deletion of gaps/missing data &#64; all positions</p></li>
</ul>
<p>Method 2:</p>
<ul class="simple">
<li><p>maximum parsimony method</p></li>
</ul>
<p>Method 3:</p>
<ul class="simple">
<li><p>UPGMA method</p></li>
</ul>
<blockquote>
<div><p>MEGA v4.0 (2007) is a program for conducting molecular evolutionary analysis of DNA sequences. Depending on the version, it packages many of the workflows used in phylogenetic analysis in one program. Particularily for v4.0, they boast the ability to generate captions, and use maximum composite likelihood (MCL) method to estimate evolutionary distances between all pairs. Recent version has expanded it’s feature set and performance. It is available for Windows, Mac and Linux. It is free for academic and non-profit use. <span id="id4">[<a class="reference internal" href="../references.html#id27" title="K. Tamura, J. Dudley, M. Nei, and S. Kumar. MEGA4: molecular evolutionary genetics analysis (MEGA) software version 4.0. Molecular Biology and Evolution, 24(8):1596–1599, apr 2007. doi:10.1093/molbev/msm092.">Tamura <em>et al.</em>, 2007</a>]</span></p>
</div></blockquote>
</li>
</ol>
<figure class="align-default" id="pike-2013-tree">
<img alt="../../_images/papers_tree.png" src="../../_images/papers_tree.png" />
<figcaption>
<p><span class="caption-number">Fig. 1 </span><span class="caption-text">Tree showing the phylogenetic relationship of Endozoicomonas to other bacterial species (43 type strains based on 16S rRNA gene sequences).</span><a class="headerlink" href="#pike-2013-tree" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>But first, we need to wrangle our data.</p>
</section>
<section id="wrangle-data">
<h2>Wrangle data<a class="headerlink" href="#wrangle-data" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="kn">import</span> <span class="nn">os</span>
</pre></div>
</div>
</div>
</div>
<p>Lets write a function that can take a list of directories and return all of the fasta sequences within them.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dir_paths</span> <span class="o">=</span> <span class="p">[</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">realpath</span><span class="p">(</span><span class="s2">&quot;../../datasets/common&quot;</span><span class="p">),</span>
             <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">realpath</span><span class="p">(</span><span class="s2">&quot;../../datasets/pike_basic_analysis&quot;</span><span class="p">)]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">parse_fasta</span><span class="p">(</span><span class="n">dir_paths</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">]:</span>
  <span class="n">fasta_records</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

  <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">dir_paths</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">dir_path</span> <span class="ow">in</span> <span class="n">dir_paths</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">path</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">dir_path</span><span class="p">):</span>
          <span class="n">file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dir_path</span><span class="p">,</span> <span class="n">path</span><span class="p">)</span>
          <span class="k">if</span> <span class="s2">&quot;fasta&quot;</span> <span class="ow">in</span> <span class="n">file</span><span class="p">:</span>
            <span class="n">fasta_record</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">file</span><span class="p">),</span> <span class="s2">&quot;fasta&quot;</span><span class="p">))</span>
            <span class="n">fasta_records</span> <span class="o">=</span> <span class="n">fasta_records</span> <span class="o">+</span> <span class="n">fasta_record</span>

  <span class="k">else</span><span class="p">:</span>
    <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;dir_paths must be a string or list of strings&quot;</span><span class="p">)</span>

  <span class="k">return</span> <span class="n">fasta_records</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">remove_duplicate</span><span class="p">(</span><span class="n">fasta_records</span><span class="p">:</span> <span class="nb">list</span><span class="p">):</span>
  <span class="n">accession_ids</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
  <span class="n">duplicates</span> <span class="o">=</span> <span class="p">[]</span>
  <span class="n">duplicate_ids</span> <span class="o">=</span> <span class="p">[]</span>

  <span class="n">retained_records</span> <span class="o">=</span> <span class="p">[]</span>
  <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">fasta</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">fasta_records</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">fasta</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">accession_ids</span><span class="p">:</span>
      <span class="n">duplicates</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
      <span class="n">duplicate_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="n">retained_records</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fasta</span><span class="p">)</span>
    <span class="n">accession_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>

  <span class="k">return</span> <span class="n">duplicates</span><span class="p">,</span> <span class="n">duplicate_ids</span><span class="p">,</span> <span class="n">retained_records</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">write_unaligned_fasta</span><span class="p">(</span><span class="n">dir_paths_for_fastas</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">id_path_str</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
  <span class="n">dir_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">id_path_str</span><span class="p">)</span>
  <span class="n">ids</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">id_path_str</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span>
  <span class="n">fasta_records</span> <span class="o">=</span> <span class="n">parse_fasta</span><span class="p">(</span><span class="n">dir_paths_for_fastas</span><span class="p">)</span>
  <span class="n">duplicates</span><span class="p">,</span> <span class="n">duplicate_ids</span><span class="p">,</span> <span class="n">retained_records</span> <span class="o">=</span> <span class="n">remove_duplicate</span><span class="p">(</span><span class="n">fasta_records</span><span class="p">)</span>
  <span class="n">filtered</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">ids</span><span class="p">,</span> <span class="n">retained_records</span><span class="p">))</span>
  <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">filtered</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dir_path</span><span class="p">,</span> <span class="s2">&quot;all_unaligned.fasta&quot;</span><span class="p">),</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
  <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Done writing unaligned fasta&quot;</span><span class="p">)</span>
  <span class="k">return</span> <span class="n">filtered</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="pike-et-al-2013-basic-analysis">
<h2>Pike et al (2013) basic analysis<a class="headerlink" href="#pike-et-al-2013-basic-analysis" title="Permalink to this heading">#</a></h2>
<p>Let us load the fasta files for the 16S sequences from the Pike et al (2013) study.</p>
<p><strong>To repeat the basic analysis, I did the following:</strong></p>
<ol class="arabic">
<li><p>I downloaded <a class="reference external" href="https://www.megasoftware.net/">MEGA but version 11</a> <span id="id5">[<a class="reference internal" href="../references.html#id29" title="Koichiro Tamura, Glen Stecher, and Sudhir Kumar. MEGA11: molecular evolutionary genetics analysis version 11. Molecular Biology and Evolution, 38(7):3022–3027, apr 2021. doi:10.1093/molbev/msab120.">Tamura <em>et al.</em>, 2021</a>]</span>.</p></li>
<li><p>I downloaded the 16S sequences using the <a class="reference download internal" download="" href="../../_downloads/81de2f88ce8f5646011e7a1fbd680aed/list_of_accession.txt"><span class="xref download myst">accession ID</span></a> in the Endozoicomonas subtree (<a class="reference internal" href="#pike-2013-tree"><span class="std std-numref">Fig. 1</span></a>).</p></li>
<li><p>I concatenated the sequences (see below).</p></li>
<li><p>I aligned the sequences using MUSCLE v5.1 <span id="id6">[<a class="reference internal" href="../references.html#id30" title="Robert C. Edgar. High-accuracy alignment ensembles enable unbiased assessments of sequence homology and phylogeny. bioRxiv, 2022. URL: https://www.biorxiv.org/content/early/2022/08/01/2021.06.20.449169, arXiv:https://www.biorxiv.org/content/early/2022/08/01/2021.06.20.449169.full.pdf, doi:10.1101/2021.06.20.449169.">Edgar, 2022</a>]</span>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>muscle5.1<span class="w"> </span>-input<span class="w"> </span>outputs/pike_basic_analysis/all_unaligned.fasta<span class="w"> </span>-output<span class="w"> </span>outputs/pike_basic_analysis/muscle_aligned.fasta
</pre></div>
</div>
</li>
<li><p>Inside MEGA I selected the following settings:</p></li>
</ol>
<table class="colwidths-auto table" id="id19">
<caption><span class="caption-number">Table 1 </span><span class="caption-text">MEGA settings to repeat basic analysis</span><a class="headerlink" href="#id19" title="Permalink to this table">#</a></caption>
<thead>
<tr class="row-odd"><th class="head"><p>Setting</p></th>
<th class="head"><p>Value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Scope</p></td>
<td><p>All taxa</p></td>
</tr>
<tr class="row-odd"><td><p>Statistical method</p></td>
<td><p>Maximum evolution</p></td>
</tr>
<tr class="row-even"><td><p>Test of phylogeny</p></td>
<td><p>Bootstrap method</p></td>
</tr>
<tr class="row-odd"><td><p>No. of bootstrap replicates</p></td>
<td><p>1000</p></td>
</tr>
<tr class="row-even"><td><p>Substitution type</p></td>
<td><p>Nucleotide</p></td>
</tr>
<tr class="row-odd"><td><p>Model/method</p></td>
<td><p>Juke-Cantor model</p></td>
</tr>
<tr class="row-even"><td><p>Gaps/missing data treatment</p></td>
<td><p>Complete deletion</p></td>
</tr>
<tr class="row-odd"><td><p>ME Search Level</p></td>
<td><p>1</p></td>
</tr>
</tbody>
</table>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">write_unaligned_fasta</span><span class="p">(</span><span class="n">dir_paths</span><span class="p">,</span> <span class="s2">&quot;../../outputs/pike_basic_analysis/list_of_accession.txt&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Done writing unaligned fasta
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[SeqRecord(seq=Seq(&#39;CTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAGCGGTAG...CGG&#39;), id=&#39;AB196667.1&#39;, name=&#39;AB196667.1&#39;, description=&#39;AB196667.1 Endozoicimonas elysicola gene for 16S ribosomal RNA, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;ATTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAGCGGTAACAGAACTAG...AAG&#39;), id=&#39;AB695088.1&#39;, name=&#39;AB695088.1&#39;, description=&#39;AB695088.1 Endozoicomonas numazuensis gene for 16S rRNA, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;TGCAAGTCGAGCGGTAACAGAACTAGCTTGCTAGTTGCTGACGAGCGGCGGACG...AAG&#39;), id=&#39;FJ347758.1&#39;, name=&#39;FJ347758.1&#39;, description=&#39;FJ347758.1 Endozoicomonas montiporae CL-33 16S ribosomal RNA gene, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;AGAGTTTTGGATCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACATGCAA...GGC&#39;), id=&#39;JX488684.2&#39;, name=&#39;JX488684.2&#39;, description=&#39;JX488684.2 Endozoicomonas euniceicola strain EF212 16S ribosomal RNA gene, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;AGAGTTTGATCCTGGCGCAGATTGAACGCTGGCGGCAGGCCTAACACATGCAAG...GTC&#39;), id=&#39;JX488685.1&#39;, name=&#39;JX488685.1&#39;, description=&#39;JX488685.1 Endozoicomonas gorgoniicola strain PS125 16S ribosomal RNA gene, partial sequence&#39;, dbxrefs=[])]
</pre></div>
</div>
</div>
</div>
</section>
<section id="pilot-analysis-1-2">
<h2>Pilot Analysis (1 &amp; 2)<a class="headerlink" href="#pilot-analysis-1-2" title="Permalink to this heading">#</a></h2>
<p>To make improvement on the basic analysis, I added more 16S sequences by searching for Endozoicomonas &amp; coral in the NCBI database. I targeted sequences that were fairly long (~1000 or more, except for one sequence) I downloaded the sequences and aligned them using MUSCLE (v.5.1) from CLI <span id="id7">[<a class="reference internal" href="../references.html#id30" title="Robert C. Edgar. High-accuracy alignment ensembles enable unbiased assessments of sequence homology and phylogeny. bioRxiv, 2022. URL: https://www.biorxiv.org/content/early/2022/08/01/2021.06.20.449169, arXiv:https://www.biorxiv.org/content/early/2022/08/01/2021.06.20.449169.full.pdf, doi:10.1101/2021.06.20.449169.">Edgar, 2022</a>]</span>.</p>
<table class="colwidths-auto table" id="id20">
<caption><span class="caption-number">Table 2 </span><span class="caption-text">List of sequences added to the alignment</span><a class="headerlink" href="#id20" title="Permalink to this table">#</a></caption>
<thead>
<tr class="row-odd"><th class="text-left head"><p>Species (or Isolate/Strain)</p></th>
<th class="text-center head"><p>Description</p></th>
<th class="text-left head"><p>Length</p></th>
<th class="text-left head"><p>Accession</p></th>
<th class="text-left head"><p>GI</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-left"><p>Endozoicomonas <strong>atrinae</strong> strain WP70</p></td>
<td class="text-center"><p>16S ribosomal RNA gene, partial sequence</p></td>
<td class="text-left"><p>1,465 bp linear DNA</p></td>
<td class="text-left"><p>KC878324.1</p></td>
<td class="text-left"><p>499141103</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>Endozoicomonas <strong>montiporae</strong> strain Ab112_MC</p></td>
<td class="text-center"><p>16S ribosomal RNA (16S) gene, complete sequence</p></td>
<td class="text-left"><p>1,024 bp linear DNA</p></td>
<td class="text-left"><p>KJ372452.1</p></td>
<td class="text-left"><p>646280569</p></td>
</tr>
<tr class="row-even"><td class="text-left"><p>Endozoicomonas <strong>ascidiicola</strong> strain AVMART05</p></td>
<td class="text-center"><p>16S ribosomal RNA gene, partial sequence</p></td>
<td class="text-left"><p>1,501 bp linear DNA</p></td>
<td class="text-left"><p>KT364257.1</p></td>
<td class="text-left"><p>1016200855</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>Endozoicomonas <strong>sp.</strong> isolate PM28062</p></td>
<td class="text-center"><p>16S ribosomal RNA gene, partial sequence</p></td>
<td class="text-left"><p>1,408 bp linear DNA</p></td>
<td class="text-left"><p>KX780138.1</p></td>
<td class="text-left"><p>1226602311</p></td>
</tr>
<tr class="row-even"><td class="text-left"><p>Uncultured Endozoicomonas <strong>sp.</strong></p></td>
<td class="text-center"><p>partial 16S rRNA gene, clone Dpd21_3_44</p></td>
<td class="text-left"><p>1,473 bp linear DNA</p></td>
<td class="text-left"><p>LN626318.1</p></td>
<td class="text-left"><p>966207698</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>Endozoicomonas <strong>sp.</strong> strain Acr-14</p></td>
<td class="text-center"><p>partial 16S rRNA gene, isolate Sea coral</p></td>
<td class="text-left"><p>1,452 bp linear DNA</p></td>
<td class="text-left"><p>LN875493.1</p></td>
<td class="text-left"><p>916534561</p></td>
</tr>
<tr class="row-even"><td class="text-left"><p>Endozoicomonas <strong>sp.</strong> Acr-12</p></td>
<td class="text-center"><p>partial 16S rRNA gene, strain Acr-12, isolate Sea coral</p></td>
<td class="text-left"><p>1,450 bp linear DNA</p></td>
<td class="text-left"><p>LN879492.1</p></td>
<td class="text-left"><p>928189784</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>Endozoicomonas <strong>sp.</strong> strain LZHN29</p></td>
<td class="text-center"><p>16S ribosomal RNA gene, partial sequence</p></td>
<td class="text-left"><p>1,550 bp linear DNA</p></td>
<td class="text-left"><p>MH201322.1</p></td>
<td class="text-left"><p>1377517605</p></td>
</tr>
<tr class="row-even"><td class="text-left"><p>Endozoicomonas <strong>sp.</strong> Hp36</p></td>
<td class="text-center"><p>16S ribosomal RNA gene, partial sequence</p></td>
<td class="text-left"><p>1,531 bp linear DNA</p></td>
<td class="text-left"><p>MK633876.1</p></td>
<td class="text-left"><p>1591471162</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>Endozoicomonas <strong>sp.</strong> strain XS200</p></td>
<td class="text-center"><p>16S ribosomal RNA gene, partial sequence</p></td>
<td class="text-left"><p>562 bp linear DNA</p></td>
<td class="text-left"><p>OQ618154.1</p></td>
<td class="text-left"><p>2456913731</p></td>
</tr>
</tbody>
</table>
<section id="pilot-1">
<h3>Pilot #1<a class="headerlink" href="#pilot-1" title="Permalink to this heading">#</a></h3>
<p>This was a run where I had one sequence missing (FJ347758.1) from the original study from <span id="id8">[<a class="reference internal" href="../references.html#id28" title="Rebecca E. Pike, Brad Haltli, and Russell G. Kerr. Description of endozoicomonas euniceicola sp. nov. and endozoicomonas gorgoniicola sp. nov., bacteria isolated from the octocorals eunicea fusca and plexaura sp., and an emended description of the genus endozoicomonas. International Journal of Systematic and Evolutionary Microbiology, 63(Pt_11):4294–4302, nov 2013. doi:10.1099/ijs.0.051490-0.">Pike <em>et al.</em>, 2013</a>]</span>.</p>
<ol class="arabic">
<li><p>I downloaded the 16S sequences using the <a class="reference download internal" download="" href="../../_downloads/c99669e7c67fd19d04d6039fca75ee1b/accession_id_list.txt"><span class="xref download myst">accession ID</span></a>.</p></li>
<li><p>I concatenated the sequences (see below).</p></li>
<li><p>I aligned the sequences using MUSCLE v5.1 <span id="id9">[<a class="reference internal" href="../references.html#id30" title="Robert C. Edgar. High-accuracy alignment ensembles enable unbiased assessments of sequence homology and phylogeny. bioRxiv, 2022. URL: https://www.biorxiv.org/content/early/2022/08/01/2021.06.20.449169, arXiv:https://www.biorxiv.org/content/early/2022/08/01/2021.06.20.449169.full.pdf, doi:10.1101/2021.06.20.449169.">Edgar, 2022</a>]</span>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>muscle5.1<span class="w"> </span>-input<span class="w"> </span>outputs/pilot_1/all_unaligned.fasta<span class="w"> </span>-output<span class="w"> </span>outputs/pilot_1/muscle/Endozoicomonas_sp.phy
</pre></div>
</div>
</li>
<li><p>I used modeltest-ng in RaxMLGUI <span id="id10">[<a class="reference internal" href="../references.html#id33" title="Diego Darriba, David Posada, Alexey M Kozlov, Alexandros Stamatakis, Benoit Morel, and Tomas Flouri. ModelTest-NG: A New and Scalable Tool for the Selection of DNA and Protein Evolutionary Models. Molecular Biology and Evolution, 37(1):291-294, 08 2019. URL: https://doi.org/10.1093/molbev/msz189, arXiv:https://academic.oup.com/mbe/article-pdf/37/1/291/32085561/msz189.pdf, doi:10.1093/molbev/msz189.">Darriba <em>et al.</em>, 2019</a>, <a class="reference internal" href="../references.html#id32" title="Daniel Edler, Johannes Klein, Alexandre Antonelli, and Daniele Silvestro. Raxmlgui 2.0: a graphical interface and toolkit for phylogenetic analyses using raxml. Methods in Ecology and Evolution, 12(2):373-377, 2021. URL: https://besjournals.onlinelibrary.wiley.com/doi/abs/10.1111/2041-210X.13512, arXiv:https://besjournals.onlinelibrary.wiley.com/doi/pdf/10.1111/2041-210X.13512, doi:https://doi.org/10.1111/2041-210X.13512.">Edler <em>et al.</em>, 2021</a>]</span>. <strong>TPM1uf+G4</strong> was the best model.</p></li>
<li><p>I then tree searched using raxmlHPC in RaxMLGUI <span id="id11">[<a class="reference internal" href="../references.html#id32" title="Daniel Edler, Johannes Klein, Alexandre Antonelli, and Daniele Silvestro. Raxmlgui 2.0: a graphical interface and toolkit for phylogenetic analyses using raxml. Methods in Ecology and Evolution, 12(2):373-377, 2021. URL: https://besjournals.onlinelibrary.wiley.com/doi/abs/10.1111/2041-210X.13512, arXiv:https://besjournals.onlinelibrary.wiley.com/doi/pdf/10.1111/2041-210X.13512, doi:https://doi.org/10.1111/2041-210X.13512.">Edler <em>et al.</em>, 2021</a>, <a class="reference internal" href="../references.html#id35" title="Alexey M Kozlov, Diego Darriba, Tomáš Flouri, Benoit Morel, and Alexandros Stamatakis. RAxML-NG: a fast, scalable and user-friendly tool for maximum likelihood phylogenetic inference. Bioinformatics, 35(21):4453-4455, 05 2019. URL: https://doi.org/10.1093/bioinformatics/btz305, arXiv:https://academic.oup.com/bioinformatics/article-pdf/35/21/4453/30330793/btz305.pdf, doi:10.1093/bioinformatics/btz305.">Kozlov <em>et al.</em>, 2019</a>]</span>. However, it seems like I made a mistake and ran it with GTRGAMMAIX instead, using ML + thorough bootstrap.</p></li>
</ol>
<p><strong>This produced:</strong></p>
<figure class="align-default" id="pilot-1-raxml-tree">
<img alt="../../_images/Ito_Kaede_PartB_screenshot_page-0001.jpg" src="../../_images/Ito_Kaede_PartB_screenshot_page-0001.jpg" />
<figcaption>
<p><span class="caption-number">Fig. 2 </span><span class="caption-text">Tree searched using ML inference with RaxMLGUI <span id="id12">[<a class="reference internal" href="../references.html#id32" title="Daniel Edler, Johannes Klein, Alexandre Antonelli, and Daniele Silvestro. Raxmlgui 2.0: a graphical interface and toolkit for phylogenetic analyses using raxml. Methods in Ecology and Evolution, 12(2):373-377, 2021. URL: https://besjournals.onlinelibrary.wiley.com/doi/abs/10.1111/2041-210X.13512, arXiv:https://besjournals.onlinelibrary.wiley.com/doi/pdf/10.1111/2041-210X.13512, doi:https://doi.org/10.1111/2041-210X.13512.">Edler <em>et al.</em>, 2021</a>, <a class="reference internal" href="../references.html#id35" title="Alexey M Kozlov, Diego Darriba, Tomáš Flouri, Benoit Morel, and Alexandros Stamatakis. RAxML-NG: a fast, scalable and user-friendly tool for maximum likelihood phylogenetic inference. Bioinformatics, 35(21):4453-4455, 05 2019. URL: https://doi.org/10.1093/bioinformatics/btz305, arXiv:https://academic.oup.com/bioinformatics/article-pdf/35/21/4453/30330793/btz305.pdf, doi:10.1093/bioinformatics/btz305.">Kozlov <em>et al.</em>, 2019</a>]</span>.</span><a class="headerlink" href="#pilot-1-raxml-tree" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">write_unaligned_fasta</span><span class="p">(</span><span class="n">dir_paths</span><span class="p">,</span> <span class="s2">&quot;../../outputs/pilot_1/accession_id_list.txt&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Done writing unaligned fasta
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[SeqRecord(seq=Seq(&#39;ATTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAGCGGTAGCGGGAAGAG...GTG&#39;), id=&#39;KC878324.1&#39;, name=&#39;KC878324.1&#39;, description=&#39;KC878324.1 Endozoicomonas atrinae strain WP70 16S ribosomal RNA gene, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;ATGCAGTCGAGCGGTGACAGAACTAGCTTGCTAGTTGCTGACGAGCGGCGGACG...AGG&#39;), id=&#39;KJ372452.1&#39;, name=&#39;KJ372452.1&#39;, description=&#39;KJ372452.1 Endozoicomonas montiporae strain Ab112_MC 16S ribosomal RNA (16S) gene, complete sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;AGAGTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACATGCAAG...GTA&#39;), id=&#39;KT364257.1&#39;, name=&#39;KT364257.1&#39;, description=&#39;KT364257.1 Endozoicomonas ascidiicola strain AVMART05 16S ribosomal RNA gene, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;TAACATTCCCAGCTTGCTGGGAGATGACGAGCGGCGGACGGGTGAGTAACACGT...GAT&#39;), id=&#39;KX780138.1&#39;, name=&#39;KX780138.1&#39;, description=&#39;KX780138.1 Endozoicomonas sp. isolate PM28062 16S ribosomal RNA gene, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;TGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAGCGGTAGC...GTG&#39;), id=&#39;LN626318.1&#39;, name=&#39;LN626318.1&#39;, description=&#39;LN626318.1 Uncultured Endozoicomonas sp. partial 16S rRNA gene, clone Dpd21_3_44&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;GTCGAGCGGTAGCAGGAAGTGCTTGCACTTCGCTGACGAGCGGCGGACGGGTGC...TCG&#39;), id=&#39;LN875493.1&#39;, name=&#39;LN875493.1&#39;, description=&#39;LN875493.1 Endozoicomonas sp. Acr-14 partial 16S rRNA gene, strain Acr-14, isolate Sea coral&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;GCAGTCGAGCGGTAGCAGGAAGTAGCTTGCACTTCGCTGACGAGCGGCGGACGG...GTG&#39;), id=&#39;LN879492.1&#39;, name=&#39;LN879492.1&#39;, description=&#39;LN879492.1 Endozoicomonas sp. Acr-12 partial 16S rRNA gene, strain Acr-12, isolate Sea coral&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;TCTAGAGATTAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGGCTTAA...TAA&#39;), id=&#39;MH201322.1&#39;, name=&#39;MH201322.1&#39;, description=&#39;MH201322.1 Endozoicomonas sp. strain LZHN29 16S ribosomal RNA gene, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;AGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACATGCAAG...CCT&#39;), id=&#39;MK633876.1&#39;, name=&#39;MK633876.1&#39;, description=&#39;MK633876.1 Endozoicomonas sp. Hp36 16S ribosomal RNA gene, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;CACGTAGGAATCTGCCTGGTAGTGGGGGATAGCCCGGAGAAATCCGGATTAATA...GAG&#39;), id=&#39;OQ618154.1&#39;, name=&#39;OQ618154.1&#39;, description=&#39;OQ618154.1 Endozoicomonas sp. strain XS200 16S ribosomal RNA gene, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;CTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAGCGGTAG...CGG&#39;), id=&#39;AB196667.1&#39;, name=&#39;AB196667.1&#39;, description=&#39;AB196667.1 Endozoicimonas elysicola gene for 16S ribosomal RNA, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;ATTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAGCGGTAACAGAACTAG...AAG&#39;), id=&#39;AB695088.1&#39;, name=&#39;AB695088.1&#39;, description=&#39;AB695088.1 Endozoicomonas numazuensis gene for 16S rRNA, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;AGAGTTTTGGATCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACATGCAA...GGC&#39;), id=&#39;JX488684.2&#39;, name=&#39;JX488684.2&#39;, description=&#39;JX488684.2 Endozoicomonas euniceicola strain EF212 16S ribosomal RNA gene, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;AGAGTTTGATCCTGGCGCAGATTGAACGCTGGCGGCAGGCCTAACACATGCAAG...GTC&#39;), id=&#39;JX488685.1&#39;, name=&#39;JX488685.1&#39;, description=&#39;JX488685.1 Endozoicomonas gorgoniicola strain PS125 16S ribosomal RNA gene, partial sequence&#39;, dbxrefs=[])]
</pre></div>
</div>
</div>
</div>
</section>
<section id="pilot-2">
<h3>Pilot #2<a class="headerlink" href="#pilot-2" title="Permalink to this heading">#</a></h3>
<p>Then I did a model test in RaxMLGUI (v2.0.10) <span id="id13">[<a class="reference internal" href="../references.html#id33" title="Diego Darriba, David Posada, Alexey M Kozlov, Alexandros Stamatakis, Benoit Morel, and Tomas Flouri. ModelTest-NG: A New and Scalable Tool for the Selection of DNA and Protein Evolutionary Models. Molecular Biology and Evolution, 37(1):291-294, 08 2019. URL: https://doi.org/10.1093/molbev/msz189, arXiv:https://academic.oup.com/mbe/article-pdf/37/1/291/32085561/msz189.pdf, doi:10.1093/molbev/msz189.">Darriba <em>et al.</em>, 2019</a>, <a class="reference internal" href="../references.html#id32" title="Daniel Edler, Johannes Klein, Alexandre Antonelli, and Daniele Silvestro. Raxmlgui 2.0: a graphical interface and toolkit for phylogenetic analyses using raxml. Methods in Ecology and Evolution, 12(2):373-377, 2021. URL: https://besjournals.onlinelibrary.wiley.com/doi/abs/10.1111/2041-210X.13512, arXiv:https://besjournals.onlinelibrary.wiley.com/doi/pdf/10.1111/2041-210X.13512, doi:https://doi.org/10.1111/2041-210X.13512.">Edler <em>et al.</em>, 2021</a>]</span> using the full list of sequences (including the one missing) which produced TrN (Tamura-Nei) as the best model.</p>
<ol class="arabic">
<li><p>I downloaded the 16S sequences using the <a class="reference download internal" download="" href="../../_downloads/f599c8c3520015e01968f92ce51e157d/accession_id_list.txt"><span class="xref download myst">accession ID</span></a>.</p></li>
<li><p>I concatenated the sequences (see below) (which included the sequences in the Endozoicomonas subtree (<a class="reference internal" href="#pike-2013-tree"><span class="std std-numref">Fig. 1</span></a>)).</p></li>
<li><p>I aligned the sequences using MUSCLE v5.1 <span id="id14">[<a class="reference internal" href="../references.html#id30" title="Robert C. Edgar. High-accuracy alignment ensembles enable unbiased assessments of sequence homology and phylogeny. bioRxiv, 2022. URL: https://www.biorxiv.org/content/early/2022/08/01/2021.06.20.449169, arXiv:https://www.biorxiv.org/content/early/2022/08/01/2021.06.20.449169.full.pdf, doi:10.1101/2021.06.20.449169.">Edgar, 2022</a>]</span>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>muscle5.1<span class="w"> </span>-input<span class="w"> </span>outputs/pilot_2/all_unaligned.fasta<span class="w"> </span>-output<span class="w"> </span>outputs/pilot_2/muscle/16S_all_aligned.fasta
</pre></div>
</div>
</li>
<li><p>Inside MEGA I selected the following settings:</p></li>
</ol>
<table class="colwidths-auto table" id="id21">
<caption><span class="caption-number">Table 3 </span><span class="caption-text">MEGA settings for basic analysis (with missing sequence)</span><a class="headerlink" href="#id21" title="Permalink to this table">#</a></caption>
<thead>
<tr class="row-odd"><th class="head"><p>Setting</p></th>
<th class="head"><p>Value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Scope</p></td>
<td><p>All taxa</p></td>
</tr>
<tr class="row-odd"><td><p>Statistical method</p></td>
<td><p>Maximum likelihood</p></td>
</tr>
<tr class="row-even"><td><p>Test of phylogeny</p></td>
<td><p>Bootstrap method</p></td>
</tr>
<tr class="row-odd"><td><p>No. of bootstrap replicates</p></td>
<td><p>700</p></td>
</tr>
<tr class="row-even"><td><p>Substitution type</p></td>
<td><p>Nucleotide</p></td>
</tr>
<tr class="row-odd"><td><p>Model/method</p></td>
<td><p>Tamura-Nei model</p></td>
</tr>
<tr class="row-even"><td><p>Gaps/missing data treatment</p></td>
<td><p>Complete deletion</p></td>
</tr>
<tr class="row-odd"><td><p>ML Heuristic method</p></td>
<td><p>Nearest-Neighbor-Interchange (NNI)</p></td>
</tr>
<tr class="row-even"><td><p>Initial tree for ML</p></td>
<td><p>Neighbor-joining</p></td>
</tr>
</tbody>
</table>
<p><strong>This produced:</strong></p>
<figure class="align-default" id="id22">
<img alt="../../_images/16s_ml_all_aligned.png" src="../../_images/16s_ml_all_aligned.png" />
<figcaption>
<p><span class="caption-number">Fig. 3 </span><span class="caption-text">The evolutionary history was inferred by using the Maximum Likelihood method and Tamura-Nei model <span id="id15">[<a class="reference internal" href="../references.html#id37" title="K Tamura and M Nei. Estimation of the number of nucleotide substitutions in the control region of mitochondrial DNA in humans and chimpanzees. Molecular Biology and Evolution, 10(3):512-526, 05 1993. URL: https://doi.org/10.1093/oxfordjournals.molbev.a040023, arXiv:https://academic.oup.com/mbe/article-pdf/10/3/512/11168171/2TAMU.PDF, doi:10.1093/oxfordjournals.molbev.a040023.">Tamura and Nei, 1993</a>]</span>. The bootstrap consensus tree inferred from 500 replicates <span id="id16">[<a class="reference internal" href="../references.html#id38" title="Joseph Felsenstein. Confidence limits on phylogenies: an approach using the bootstrap. Evolution, 39(4):783, jul 1985. doi:10.2307/2408678.">Felsenstein, 1985</a>]</span> is taken to represent the evolutionary history of the taxa analyzed. Branches corresponding to partitions reproduced in less than 50% bootstrap replicates are collapsed. The percentage of replicate trees in which the associated taxa clustered together in the bootstrap test 500 replicates are shown next to the branches <span id="id17">[<a class="reference internal" href="../references.html#id38" title="Joseph Felsenstein. Confidence limits on phylogenies: an approach using the bootstrap. Evolution, 39(4):783, jul 1985. doi:10.2307/2408678.">Felsenstein, 1985</a>]</span>. Initial tree(s) for the heuristic search were obtained by applying the Neighbor-Joining method to a matrix of pairwise distances estimated using the Tamura-Nei model. This analysis involved 20 nucleotide sequences. Codon positions included were 1st+2nd+3rd+Noncoding. All positions containing gaps and missing data were eliminated (complete deletion option). There were a total of 545 positions in the final dataset. Evolutionary analyses were conducted in MEGA11 <span id="id18">[<a class="reference internal" href="../references.html#id29" title="Koichiro Tamura, Glen Stecher, and Sudhir Kumar. MEGA11: molecular evolutionary genetics analysis version 11. Molecular Biology and Evolution, 38(7):3022–3027, apr 2021. doi:10.1093/molbev/msab120.">Tamura <em>et al.</em>, 2021</a>]</span></span><a class="headerlink" href="#id22" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">write_unaligned_fasta</span><span class="p">(</span><span class="n">dir_paths</span><span class="p">,</span> <span class="s2">&quot;../../outputs/pilot_2/accession_id_list.txt&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Done writing unaligned fasta
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[SeqRecord(seq=Seq(&#39;ATTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAGCGGTAGCGGGAAGAG...GTG&#39;), id=&#39;KC878324.1&#39;, name=&#39;KC878324.1&#39;, description=&#39;KC878324.1 Endozoicomonas atrinae strain WP70 16S ribosomal RNA gene, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;ATGCAGTCGAGCGGTGACAGAACTAGCTTGCTAGTTGCTGACGAGCGGCGGACG...AGG&#39;), id=&#39;KJ372452.1&#39;, name=&#39;KJ372452.1&#39;, description=&#39;KJ372452.1 Endozoicomonas montiporae strain Ab112_MC 16S ribosomal RNA (16S) gene, complete sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;AGAGTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACATGCAAG...GTA&#39;), id=&#39;KT364257.1&#39;, name=&#39;KT364257.1&#39;, description=&#39;KT364257.1 Endozoicomonas ascidiicola strain AVMART05 16S ribosomal RNA gene, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;TAACATTCCCAGCTTGCTGGGAGATGACGAGCGGCGGACGGGTGAGTAACACGT...GAT&#39;), id=&#39;KX780138.1&#39;, name=&#39;KX780138.1&#39;, description=&#39;KX780138.1 Endozoicomonas sp. isolate PM28062 16S ribosomal RNA gene, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;TGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAGCGGTAGC...GTG&#39;), id=&#39;LN626318.1&#39;, name=&#39;LN626318.1&#39;, description=&#39;LN626318.1 Uncultured Endozoicomonas sp. partial 16S rRNA gene, clone Dpd21_3_44&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;GTCGAGCGGTAGCAGGAAGTGCTTGCACTTCGCTGACGAGCGGCGGACGGGTGC...TCG&#39;), id=&#39;LN875493.1&#39;, name=&#39;LN875493.1&#39;, description=&#39;LN875493.1 Endozoicomonas sp. Acr-14 partial 16S rRNA gene, strain Acr-14, isolate Sea coral&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;GCAGTCGAGCGGTAGCAGGAAGTAGCTTGCACTTCGCTGACGAGCGGCGGACGG...GTG&#39;), id=&#39;LN879492.1&#39;, name=&#39;LN879492.1&#39;, description=&#39;LN879492.1 Endozoicomonas sp. Acr-12 partial 16S rRNA gene, strain Acr-12, isolate Sea coral&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;TCTAGAGATTAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGGCTTAA...TAA&#39;), id=&#39;MH201322.1&#39;, name=&#39;MH201322.1&#39;, description=&#39;MH201322.1 Endozoicomonas sp. strain LZHN29 16S ribosomal RNA gene, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;AGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACATGCAAG...CCT&#39;), id=&#39;MK633876.1&#39;, name=&#39;MK633876.1&#39;, description=&#39;MK633876.1 Endozoicomonas sp. Hp36 16S ribosomal RNA gene, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;CACGTAGGAATCTGCCTGGTAGTGGGGGATAGCCCGGAGAAATCCGGATTAATA...GAG&#39;), id=&#39;OQ618154.1&#39;, name=&#39;OQ618154.1&#39;, description=&#39;OQ618154.1 Endozoicomonas sp. strain XS200 16S ribosomal RNA gene, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;CTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAGCGGTAG...CGG&#39;), id=&#39;AB196667.1&#39;, name=&#39;AB196667.1&#39;, description=&#39;AB196667.1 Endozoicimonas elysicola gene for 16S ribosomal RNA, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;ATTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAGCGGTAACAGAACTAG...AAG&#39;), id=&#39;AB695088.1&#39;, name=&#39;AB695088.1&#39;, description=&#39;AB695088.1 Endozoicomonas numazuensis gene for 16S rRNA, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;TGCAAGTCGAGCGGTAACAGAACTAGCTTGCTAGTTGCTGACGAGCGGCGGACG...AAG&#39;), id=&#39;FJ347758.1&#39;, name=&#39;FJ347758.1&#39;, description=&#39;FJ347758.1 Endozoicomonas montiporae CL-33 16S ribosomal RNA gene, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;AGAGTTTTGGATCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACATGCAA...GGC&#39;), id=&#39;JX488684.2&#39;, name=&#39;JX488684.2&#39;, description=&#39;JX488684.2 Endozoicomonas euniceicola strain EF212 16S ribosomal RNA gene, partial sequence&#39;, dbxrefs=[]),
 SeqRecord(seq=Seq(&#39;AGAGTTTGATCCTGGCGCAGATTGAACGCTGGCGGCAGGCCTAACACATGCAAG...GTC&#39;), id=&#39;JX488685.1&#39;, name=&#39;JX488685.1&#39;, description=&#39;JX488685.1 Endozoicomonas gorgoniicola strain PS125 16S ribosomal RNA gene, partial sequence&#39;, dbxrefs=[])]
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./src\3_pilot_1_2"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  <!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="../2_proposal/literature.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Literature Review</p>
      </div>
    </a>
    <a class="right-next"
       href="../3_pilot_3/download.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Pilot 3</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#basic-analysis">Basic analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#wrangle-data">Wrangle data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pike-et-al-2013-basic-analysis">Pike et al (2013) basic analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pilot-analysis-1-2">Pilot Analysis (1 &amp; 2)</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pilot-1">Pilot #1</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pilot-2">Pilot #2</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Kaede Ito
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=ac02cc09edc035673794"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=ac02cc09edc035673794"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>