

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Final analysis download &#8212; Coral Bleaching - BIOL 417</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=ac02cc09edc035673794" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=ac02cc09edc035673794" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=ac02cc09edc035673794" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=ac02cc09edc035673794" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/drawio.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=ac02cc09edc035673794" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=ac02cc09edc035673794" />
  <script src="../../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=ac02cc09edc035673794"></script>

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'src/4_full_analysis/download';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Download genomes" href="download_genome.html" />
    <link rel="prev" title="Tree finding" href="../3_pilot_3/tree_finding.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../readme.html">
  
  
  
  
  
    <p class="title logo__title">Coral Bleaching - BIOL 417</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../readme.html">
                    Welcome to BIOL 417 101 Capstone!
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Exploring Literature</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../1_exploratory/index.html">Proposal ideas</a></li>
<li class="toctree-l1"><a class="reference internal" href="../1_exploratory/datasets.html">List of used datasets for analysis</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Proposal</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../2_proposal/index.html">Proposal</a></li>
<li class="toctree-l1"><a class="reference internal" href="../2_proposal/literature.html">Literature Review</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Pilot analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../3_pilot_1_2/index.html">Pilot 1 &amp; 2</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../3_pilot_3/download.html">Pilot 3</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../3_pilot_3/tree_finding.html">Tree finding</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Full analysis</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Final analysis download</a></li>
<li class="toctree-l1"><a class="reference internal" href="download_genome.html">Download genomes</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">End</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../references.html">References</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/Mapleia/BIOL417_Coral" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/src/4_full_analysis/download.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Final analysis download</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#setup-the-environment">Setup the environment</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#clean-and-wrangle-tables">Clean and wrangle tables</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#ncbi-popset">NCBI: Popset</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#search-in-ncbi-popset">Search in NCBI (Popset)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#summary-of-popset">Summary of popset</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#geographic-details">Geographic details</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#ncbi-nucleotide-databse">NCBI: Nucleotide databse</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#search-in-ncbi-nucleotide">Search in NCBI (Nucleotide)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#summary-of-nucleotide-database">Summary of nucleotide database</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id36">Geographic details</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#fixing-isolation-source-column">Fixing isolation_source column</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="final-analysis-download">
<h1>Final analysis download<a class="headerlink" href="#final-analysis-download" title="Permalink to this heading">#</a></h1>
<p>Since 16S shows more promising results of identifying phylogeny of bacterial taxa, I will download a new set, but specifically look at entries under “popset” first. If there aren’t enough sequences, I will expand the scope to include more sequences.</p>
<section id="setup-the-environment">
<h2>Setup the environment<a class="headerlink" href="#setup-the-environment" title="Permalink to this heading">#</a></h2>
<p>We will utilize the package <a class="reference external" href="https://biopython.org/wiki/Documentation"><code class="docutils literal notranslate"><span class="pre">BioPython</span></code></a> <span id="id1">[<a class="reference internal" href="../references.html#id36" title="Peter J. A. Cock, Tiago Antao, Jeffrey T. Chang, Brad A. Chapman, Cymon J. Cox, Andrew Dalke, Iddo Friedberg, Thomas Hamelryck, Frank Kauff, Bartek Wilczynski, and Michiel J. L. de Hoon. Biopython: freely available python tools for computational molecular biology and bioinformatics. Bioinformatics, 25(11):1422–1423, mar 2009. doi:10.1093/bioinformatics/btp163.">Cock <em>et al.</em>, 2009</a>]</span> to interface with sequencial data hosted on NCBI. We will also use <code class="docutils literal notranslate"><span class="pre">pandas</span></code> <span id="id2">[<a class="reference internal" href="../references.html#id45" title="The pandas development team. Pandas-dev/pandas: pandas. November 2023. URL: https://doi.org/10.5281/zenodo.10107975, doi:10.5281/zenodo.10107975.">pandas development team, 2023</a>, <a class="reference internal" href="../references.html#id46" title="Wes McKinney. Data Structures for Statistical Computing in Python. In Stéfan van der Walt and Jarrod Millman, editors, Proceedings of the 9th Python in Science Conference, 56 - 61. 2010. doi:10.25080/Majora-92bf1922-00a.">Wes McKinney, 2010</a>]</span> to store metadata in a dataframe.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span><span class="p">,</span> <span class="n">SeqIO</span>
<span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">Reference</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">Bio.Entrez.Parser</span> <span class="kn">import</span> <span class="n">DictionaryElement</span>
<span class="kn">from</span> <span class="nn">pandas</span> <span class="kn">import</span> <span class="n">DataFrame</span><span class="p">,</span> <span class="n">notna</span>
<span class="kn">from</span> <span class="nn">Bio.Entrez</span> <span class="kn">import</span> <span class="n">HTTPError</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">pandas</span> <span class="kn">import</span> <span class="n">merge</span>
<span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">UndefinedSequenceError</span>
<span class="kn">import</span> <span class="nn">re</span>


<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;kaedeito@student.ubc.ca&quot;</span>
<span class="c1"># This line sets the name of the tool that is making the queries</span>
<span class="n">Entrez</span><span class="o">.</span><span class="n">tool</span> <span class="o">=</span> <span class="s2">&quot;download_16s.ipynb&quot;</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s2">&quot;download_16s&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Create a folder to save our fasta and genbank file.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dir_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">realpath</span><span class="p">(</span><span class="s2">&quot;..</span><span class="se">\\</span><span class="s2">..</span><span class="se">\\</span><span class="s2">datasets</span><span class="se">\\</span><span class="s2">full_analysis&quot;</span><span class="p">)</span>
<span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">dir_path</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="clean-and-wrangle-tables">
<h2>Clean and wrangle tables<a class="headerlink" href="#clean-and-wrangle-tables" title="Permalink to this heading">#</a></h2>
<p>We need to clean the data to make it easier to work with.
We need to load the metadata into a tabular format to make it easier to find problems, or to find interesting patterns.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">clean_table</span><span class="p">(</span><span class="n">df</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">,</span> <span class="n">granularity</span><span class="p">:</span><span class="nb">int</span><span class="p">):</span>
<span class="w">  </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">  Clean the dataframe by removing rows and cleaning string formatting.</span>

<span class="sd">  :param `df`: The dataframe to clean.</span>
<span class="sd">  :param `granularity`: The number of characters to group the ids by.</span>
<span class="sd">  &quot;&quot;&quot;</span>
  <span class="n">feat_df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
  <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">feat_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>

  <span class="c1"># Split the country column into two columns</span>
  <span class="n">feat_df</span><span class="p">[[</span><span class="s1">&#39;country&#39;</span><span class="p">,</span><span class="s1">&#39;iso_source2&#39;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;country&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">,</span><span class="n">expand</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
  <span class="c1"># Trim whitespace from the isolation_source column</span>
  <span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;iso_source2&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;iso_source2&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">x</span><span class="p">)</span>

  <span class="c1"># Group the ids by the first n characters</span>
  <span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;group&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="p">[:</span><span class="n">granularity</span><span class="p">]</span>

  <span class="n">contains_coral</span> <span class="o">=</span> <span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;isolation_source&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;coral&#39;</span><span class="p">)</span>
  <span class="n">contains_sponge</span> <span class="o">=</span> <span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;isolation_source&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;sponge&#39;</span><span class="p">)</span>
  <span class="n">feat_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="n">contains_coral</span> <span class="o">|</span> <span class="n">contains_sponge</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;host&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">()),</span> <span class="s1">&#39;host&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;isolation_source&#39;</span><span class="p">]</span>
  <span class="n">feat_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="n">contains_coral</span> <span class="o">|</span> <span class="n">contains_sponge</span><span class="p">),</span> <span class="s1">&#39;isolation_source&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>

  <span class="c1"># Fill the isolation_source column with the iso_source2 column if it is null</span>
  <span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;isolation_source&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;isolation_source&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;iso_source2&#39;</span><span class="p">])</span>

  <span class="n">feat_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;group&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;KC668&#39;</span><span class="p">,</span> <span class="s1">&#39;country&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Saudi Arabia&#39;</span>
  <span class="n">feat_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;group&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;KC668&#39;</span><span class="p">,</span> <span class="s1">&#39;isolation_source&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Southern Red Sea&#39;</span>

  <span class="c1"># Clean the table by removing the rows that do not contain Endozoicomonas</span>
  <span class="n">feat_df</span> <span class="o">=</span> <span class="n">feat_df</span><span class="p">[</span><span class="n">notna</span><span class="p">(</span><span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;organism&#39;</span><span class="p">])</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;organism&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;Endozoicomonas&#39;</span><span class="p">)</span> <span class="o">|</span> <span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;organism&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;uncultured bacterium&#39;</span><span class="p">))]</span>

  <span class="c1"># Clean the table by removing the rows that do not contain 16S</span>
  <span class="n">feat_df</span> <span class="o">=</span> <span class="n">feat_df</span><span class="p">[</span><span class="n">notna</span><span class="p">(</span><span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;product&#39;</span><span class="p">])</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;product&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;16S&#39;</span><span class="p">))]</span>

  <span class="c1"># Copy the colony info from the isolation_source column</span>
  <span class="k">if</span> <span class="s1">&#39;colony&#39;</span> <span class="ow">in</span> <span class="n">feat_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
    <span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;strain&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;strain&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;colony&#39;</span><span class="p">])</span>

  <span class="k">if</span> <span class="s1">&#39;BioProject&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">feat_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
    <span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;BioProject&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>

  <span class="c1"># Split the db_xref column into multiple columns, and use the first column as column name</span>
  <span class="n">feat_df_nuc_xref</span> <span class="o">=</span> <span class="n">feat_df</span><span class="p">[[</span><span class="s1">&#39;id&#39;</span><span class="p">,</span> <span class="s1">&#39;db_xref&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
  <span class="n">feat_df_nuc_xref</span><span class="p">[[</span><span class="s1">&#39;xref&#39;</span><span class="p">,</span><span class="s1">&#39;value&#39;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">feat_df</span><span class="p">[</span><span class="s1">&#39;db_xref&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">,</span><span class="n">expand</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

  <span class="n">feat_df_nuc_xref</span> <span class="o">=</span> <span class="n">feat_df_nuc_xref</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
  <span class="n">feat_df_nuc_xref</span> <span class="o">=</span> <span class="n">feat_df_nuc_xref</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">],</span> <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;xref&#39;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;value&#39;</span><span class="p">)</span>
  <span class="n">feat_df</span> <span class="o">=</span> <span class="n">merge</span><span class="p">(</span><span class="n">feat_df</span><span class="p">,</span> <span class="n">feat_df_nuc_xref</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;id&quot;</span><span class="p">)</span>

  <span class="c1"># remove the columns that are not needed</span>
  <span class="n">feat_df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;db_xref&#39;</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

  <span class="c1"># Reset the index</span>
  <span class="n">feat_df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>

  <span class="k">return</span> <span class="n">feat_df</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">save_file</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="n">records</span><span class="p">,</span> <span class="n">rec_type</span><span class="p">):</span>
<span class="w">  </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">  Save the records to the file.</span>

<span class="sd">  `file_name`: The name of the file to save to.</span>

<span class="sd">  `records`: The records to save.</span>

<span class="sd">  `rec_type`: The type of record to save.</span>
<span class="sd">  &quot;&quot;&quot;</span>
  <span class="n">file_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dir_path</span><span class="p">,</span> <span class="n">file_name</span><span class="p">)</span>
  <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_handle</span><span class="p">:</span>
      <span class="k">try</span><span class="p">:</span>
        <span class="n">count</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="n">out_handle</span><span class="p">,</span> <span class="n">rec_type</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">rec_type</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">count</span><span class="si">}</span><span class="s2"> Saved&quot;</span><span class="p">)</span>
      <span class="k">except</span> <span class="n">UndefinedSequenceError</span> <span class="k">as</span> <span class="n">e_seq</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e_seq</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to write </span><span class="si">{</span><span class="n">rec_type</span><span class="si">}</span><span class="s2"> to file </span><span class="si">{</span><span class="n">file_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
      <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to write </span><span class="si">{</span><span class="n">rec_type</span><span class="si">}</span><span class="s2"> to file </span><span class="si">{</span><span class="n">file_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">table_features</span><span class="p">(</span><span class="n">count</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">records</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">],</span> <span class="n">rec_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">granularity</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">debug</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
<span class="w">  </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">  1. Extracts the features (both source and rRNA)</span>
<span class="sd">  2. Based on cleaned up list of features, filter records for matching ids</span>
<span class="sd">  3. Save the records to a file</span>
<span class="sd">  &quot;&quot;&quot;</span>
  <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="n">count</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">rec_type</span><span class="si">}</span><span class="s2">, sequences: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">records</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
  <span class="n">qualifiers</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[]</span>

  <span class="n">save_records</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

  <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
    <span class="n">feats</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">SeqFeature</span><span class="p">]</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span>

    <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="s1">&#39;KC668564.1&#39;</span><span class="p">:</span>
      <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Printing for KC668564.1&quot;</span><span class="p">)</span>
      <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">)</span>
      <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">feats</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>
    <span class="p">[</span><span class="n">source</span><span class="p">]</span> <span class="o">=</span> <span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">f</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;source&quot;</span><span class="p">,</span> <span class="n">feats</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">source</span><span class="p">:</span>
       <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;No source feature&quot;</span><span class="p">)</span>

    <span class="c1"># Copy the qualifiers and add the id</span>
    <span class="c1"># merge the rRNA qualifiers to the same dict</span>
    <span class="n">qual</span> <span class="o">=</span> <span class="n">source</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">qual</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span>

    <span class="c1"># Filter for the rRNA features</span>
    <span class="n">rRNA_filtered</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">f</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;rRNA&quot;</span><span class="p">,</span> <span class="n">feats</span><span class="p">))</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">rRNA_filtered</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
       <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;No rRNA feature&quot;</span><span class="p">)</span>
       <span class="k">continue</span>
    <span class="k">elif</span> <span class="s1">&#39;contig&#39;</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">:</span>
      <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Is a contig&quot;</span><span class="p">)</span>
      <span class="k">continue</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="k">for</span> <span class="n">rRNA</span> <span class="ow">in</span> <span class="n">rRNA_filtered</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
          <span class="n">qual</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">rRNA</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
           <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">rRNA</span><span class="p">)</span>
           <span class="n">logger</span><span class="o">.</span><span class="n">exception</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>

      <span class="n">reference_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Reference</span><span class="p">]</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;references&#39;</span><span class="p">,</span> <span class="p">[])</span>
      <span class="c1"># Get the pubmed id list from the references</span>
      <span class="n">pubmed_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">ref</span><span class="o">.</span><span class="n">pubmed_id</span> <span class="k">for</span> <span class="n">ref</span> <span class="ow">in</span> <span class="n">reference_list</span> <span class="k">if</span> <span class="n">ref</span><span class="o">.</span><span class="n">pubmed_id</span><span class="p">]</span>
      <span class="n">qual</span><span class="p">[</span><span class="s1">&#39;pubmed_ids&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pubmed_ids</span>

      <span class="c1"># Split the dbxref into a dict</span>
      <span class="k">for</span> <span class="n">dbxref</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">dbxrefs</span><span class="p">:</span>
        <span class="n">split</span> <span class="o">=</span> <span class="n">dbxref</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">split</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
          <span class="n">qual</span><span class="p">[</span><span class="n">split</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">split</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

      <span class="c1"># Since all feature values are a list of string, (usually of length 1), convert to string</span>
      <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">qual</span><span class="p">:</span>
        <span class="n">qual_list</span> <span class="o">=</span> <span class="n">qual</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">qual_list</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
          <span class="n">qual</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">qual_list</span><span class="p">)</span>

      <span class="c1"># Extract the colony info from the isolation_source</span>
      <span class="k">if</span> <span class="s2">&quot;Coral-associated microbial aggregate&quot;</span> <span class="ow">in</span> <span class="n">qual</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;isolation_source&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;colony info found for </span><span class="si">{</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">iso_qual</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">qual</span><span class="p">[</span><span class="s1">&#39;isolation_source&#39;</span><span class="p">]</span>

        <span class="p">[</span><span class="n">res</span><span class="p">]</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;\((.*?)\)&#39;</span><span class="p">,</span> <span class="n">iso_qual</span><span class="p">)</span>

        <span class="n">qual</span><span class="p">[</span><span class="s1">&#39;isolation_source&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">qual</span><span class="p">[</span><span class="s1">&#39;colony&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">res</span>

      <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="s1">&#39;KC668564.1&#39;</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Printing for KC668564.1&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">qual</span><span class="p">)</span>

      <span class="c1"># As a safe guard, check that the sequence is not empty</span>
      <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">:</span>
        <span class="n">save_records</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>

      <span class="n">qualifiers</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">qual</span><span class="p">)</span>

  <span class="c1"># Load features into a dataframe</span>
  <span class="n">df1</span> <span class="o">=</span> <span class="n">DataFrame</span><span class="p">(</span><span class="n">qualifiers</span><span class="p">)</span>
  <span class="c1"># Clean the table by replacing &quot;None&quot; and &quot;nan&quot;</span>
  <span class="n">df1</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;None&quot;</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
  <span class="n">df1</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">NaN</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
  <span class="n">df1</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;nan&quot;</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">regex</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

  <span class="c1"># If not in debug mode, clean the table + save the files</span>
  <span class="k">if</span> <span class="n">debug</span><span class="p">:</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">df1</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
    <span class="k">return</span> <span class="n">df1</span><span class="p">,</span> <span class="n">save_records</span>
  <span class="k">else</span><span class="p">:</span>
    <span class="n">df2</span> <span class="o">=</span> <span class="n">clean_table</span><span class="p">(</span><span class="n">df1</span><span class="p">,</span> <span class="n">granularity</span><span class="p">)</span>
    <span class="c1"># Filter the records by the ids in the dataframe</span>
    <span class="n">filtered_recs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">filter</span><span class="p">((</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">df2</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()),</span> <span class="n">save_records</span><span class="p">))</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Records to save: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">filtered_recs</span><span class="p">)</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">save_records</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Save the records to a file</span>
    <span class="n">save_file</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">rec_type</span><span class="si">}</span><span class="s2">_nucleotide.gb&quot;</span><span class="p">,</span> <span class="n">filtered_recs</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">)</span>
    <span class="n">save_file</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">rec_type</span><span class="si">}</span><span class="s2">_nucleotide.fasta&quot;</span><span class="p">,</span> <span class="n">filtered_recs</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">df2</span><span class="p">,</span> <span class="n">filtered_recs</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="ncbi-popset">
<h2>NCBI: Popset<a class="headerlink" href="#ncbi-popset" title="Permalink to this heading">#</a></h2>
<section id="search-in-ncbi-popset">
<h3>Search in NCBI (Popset)<a class="headerlink" href="#search-in-ncbi-popset" title="Permalink to this heading">#</a></h3>
<p>As noted by the NCBI’s website, their <code class="docutils literal notranslate"><span class="pre">Popset</span> <span class="pre">database</span></code> is a collection of related sequences that is sourced from a single population/phylogenetic/mutation/ecosystem study.</p>
<p>This is a useful grouping, as we would be able to extract the locational data from the metadata and use it to create a map of the distribution of the bacteria.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">search_popset</span><span class="p">(</span><span class="n">additional_ids</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]):</span>
<span class="w">  </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">  Search for the popset database for the coral and Endozoicomonas.</span>
<span class="sd">  :return: The number of results and the list of ids.</span>
<span class="sd">  &quot;&quot;&quot;</span>
  <span class="k">try</span><span class="p">:</span>
    <span class="n">term</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;((Endozoicomonas[Organism] OR Endozoicomonas[All Fields]) AND coral[All Fields]&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">term</span><span class="p">)</span>
    <span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;popset&quot;</span><span class="p">,</span>
      <span class="n">term</span><span class="o">=</span><span class="n">term</span><span class="p">,</span>
      <span class="n">retmax</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">DictionaryElement</span><span class="p">):</span>
      <span class="n">ids</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="s2">&quot;IdList&quot;</span><span class="p">]</span>
      <span class="n">ids</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">additional_ids</span><span class="p">)</span>
      <span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;popset&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">ids</span><span class="p">,</span> <span class="n">rettype</span><span class="o">=</span><span class="s2">&quot;gb&quot;</span><span class="p">)</span>
      <span class="n">records</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="s2">&quot;gb&quot;</span><span class="p">))</span>
      <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="n">records</span><span class="p">),</span> <span class="n">ids</span><span class="p">,</span> <span class="n">records</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="k">return</span> <span class="mi">0</span><span class="p">,</span> <span class="p">[],</span> <span class="p">[]</span>
  <span class="k">except</span> <span class="n">HTTPError</span> <span class="k">as</span> <span class="n">http_e</span><span class="p">:</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">http_e</span><span class="p">))</span>
    <span class="k">return</span> <span class="mi">0</span><span class="p">,</span> <span class="p">[],</span> <span class="p">[]</span>
  <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
    <span class="k">return</span> <span class="mi">0</span><span class="p">,</span> <span class="p">[],</span> <span class="p">[]</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="summary-of-popset">
<h3>Summary of popset<a class="headerlink" href="#summary-of-popset" title="Permalink to this heading">#</a></h3>
<p>After review of literature that happened in the inspection by Nucleotide DB, I found some popset that were worth adding manually.</p>
<p>This includes the popset coming out of the work of <span id="id3">[<a class="reference internal" href="../references.html#id61" title="Mark D. Speck and Stuart P. Donachie. Widespread oceanospirillaceae bacteria in$\less $i$\greater $porites$\less $/i$\greater $spp. Journal of Marine Biology, 2012:1–7, 2012. doi:10.1155/2012/746720.">Speck and Donachie, 2012</a>]</span> (<code class="docutils literal notranslate"><span class="pre">&quot;227461503&quot;</span></code>), and the work of <span id="id4">[<a class="reference internal" href="../references.html#id47" title="Till Bayer, Matthew J. Neave, Areej Alsheikh-Hussain, Manuel Aranda, Lauren K. Yum, Tracy Mincer, Konrad Hughen, Amy Apprill, and Christian R. Voolstra. The microbiome of the red sea coral stylophora pistillata is dominated by tissue-associated endozoicomonas bacteria. Applied and Environmental Microbiology, 79(15):4759–4762, aug 2013. doi:10.1128/aem.00695-13.">Bayer <em>et al.</em>, 2013</a>]</span> (<code class="docutils literal notranslate"><span class="pre">&quot;510829312&quot;</span></code>).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">additional_ids_pop</span> <span class="o">=</span> <span class="p">[</span>
  <span class="s1">&#39;227461503&#39;</span><span class="p">,</span>
  <span class="s1">&#39;510829312&#39;</span>
<span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">count_pop</span><span class="p">,</span> <span class="n">idlist_pop</span><span class="p">,</span> <span class="n">records_pop</span> <span class="o">=</span> <span class="n">search_popset</span><span class="p">(</span><span class="n">additional_ids_pop</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load the metadata, and save the records to a file</span>
<span class="n">feat_df_pop0</span><span class="p">,</span> <span class="n">recs_pop</span> <span class="o">=</span> <span class="n">table_features</span><span class="p">(</span><span class="n">count_pop</span><span class="p">,</span> <span class="n">records_pop</span><span class="p">,</span> <span class="s2">&quot;popset&quot;</span><span class="p">)</span>

<span class="n">feat_df_pop0</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dir_path</span><span class="p">,</span> <span class="s2">&quot;features_of_popset_seq.csv&quot;</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">feat_df_pop</span> <span class="o">=</span> <span class="n">feat_df_pop0</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">col_interest</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">,</span> <span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="s1">&#39;organism&#39;</span><span class="p">,</span> <span class="s1">&#39;strain&#39;</span><span class="p">,</span> <span class="s1">&#39;host&#39;</span><span class="p">,</span> <span class="s1">&#39;country&#39;</span><span class="p">,</span> <span class="s1">&#39;taxon&#39;</span><span class="p">,</span> <span class="s1">&#39;BioProject&#39;</span><span class="p">,</span> <span class="s1">&#39;pubmed_ids&#39;</span><span class="p">,</span> <span class="s1">&#39;isolation_source&#39;</span><span class="p">,</span> <span class="s1">&#39;product&#39;</span><span class="p">]</span>
<span class="n">feat_df_pop</span> <span class="o">=</span> <span class="n">feat_df_pop</span><span class="p">[</span><span class="n">col_interest</span><span class="p">]</span>
<span class="n">feat_df_pop</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO:download_16s:Records to save: 2640/2930
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>id</th>
      <th>group</th>
      <th>organism</th>
      <th>strain</th>
      <th>host</th>
      <th>country</th>
      <th>taxon</th>
      <th>BioProject</th>
      <th>pubmed_ids</th>
      <th>isolation_source</th>
      <th>product</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>JX488684.2</td>
      <td>JX488</td>
      <td>Endozoicomonas euniceicola</td>
      <td>EF212</td>
      <td>Eunicea fusca (gorgonian coral)</td>
      <td>USA</td>
      <td>1234143</td>
      <td>NaN</td>
      <td>23832969</td>
      <td>Hillsboro Ledge</td>
      <td>16S ribosomal RNA</td>
    </tr>
    <tr>
      <th>1</th>
      <td>JX488685.1</td>
      <td>JX488</td>
      <td>Endozoicomonas gorgoniicola</td>
      <td>PS125</td>
      <td>Plexaura sp. (gorgonian coral)</td>
      <td>Bahamas</td>
      <td>1234144</td>
      <td>NaN</td>
      <td>23832969</td>
      <td>Tuna Alley (inside reef)</td>
      <td>16S ribosomal RNA</td>
    </tr>
    <tr>
      <th>2</th>
      <td>MG725756.1</td>
      <td>MG725</td>
      <td>uncultured bacterium</td>
      <td>NaN</td>
      <td>Pocillopora verrucosa</td>
      <td>NaN</td>
      <td>77133</td>
      <td>NaN</td>
      <td>29468040</td>
      <td>NaN</td>
      <td>16S ribosomal RNA</td>
    </tr>
    <tr>
      <th>3</th>
      <td>MG725757.1</td>
      <td>MG725</td>
      <td>uncultured bacterium</td>
      <td>NaN</td>
      <td>Pocillopora verrucosa</td>
      <td>NaN</td>
      <td>77133</td>
      <td>NaN</td>
      <td>29468040</td>
      <td>NaN</td>
      <td>16S ribosomal RNA</td>
    </tr>
    <tr>
      <th>4</th>
      <td>MG725758.1</td>
      <td>MG725</td>
      <td>uncultured bacterium</td>
      <td>NaN</td>
      <td>Pocillopora verrucosa</td>
      <td>NaN</td>
      <td>77133</td>
      <td>NaN</td>
      <td>29468040</td>
      <td>NaN</td>
      <td>16S ribosomal RNA</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="geographic-details">
<h3>Geographic details<a class="headerlink" href="#geographic-details" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">feat_df_pop</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;country&#39;</span><span class="p">],</span> <span class="n">dropna</span><span class="o">=</span><span class="kc">False</span><span class="p">)[</span><span class="s1">&#39;country&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>country
Bahamas            1
France             1
Malaysia           5
Saudi Arabia    1172
Taiwan            39
USA              784
NaN                0
Name: country, dtype: int64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">feat_df_pop</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;country&#39;</span><span class="p">,</span> <span class="s1">&#39;isolation_source&#39;</span><span class="p">],</span> <span class="n">dropna</span><span class="o">=</span><span class="kc">False</span><span class="p">)[</span><span class="s1">&#39;isolation_source&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>country       isolation_source        
Bahamas       Tuna Alley (inside reef)       1
France        NaN                            0
Malaysia      NaN                            0
Saudi Arabia  Southern Red Sea            1172
Taiwan        Green Island                  17
              Ken Ting                      21
              Yeh Liu                        1
USA           Hillsboro Ledge                1
              Summerland Key, Florida       25
              West Maui                    758
NaN           NaN                            0
Name: isolation_source, dtype: int64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">feat_df_pop</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;host&#39;</span><span class="p">,</span> <span class="s1">&#39;country&#39;</span><span class="p">,</span> <span class="s1">&#39;isolation_source&#39;</span><span class="p">],</span> <span class="n">dropna</span><span class="o">=</span><span class="kc">False</span><span class="p">)[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>host                                  country       isolation_source        
Acropora clathrata                    NaN           NaN                           1
Acropora humilis                      Saudi Arabia  Southern Red Sea            322
Corallium rubrum                      France        NaN                           1
Eunicea fusca (gorgonian coral)       USA           Hillsboro Ledge               1
Eunicella cavolini                    NaN           NaN                          20
Lobophytum compactum                  NaN           NaN                           1
Marine tunicates                      Malaysia      NaN                           5
Plexaura homomalla (gorgonian coral)  USA           Summerland Key, Florida      25
Plexaura sp. (gorgonian coral)        Bahamas       Tuna Alley (inside reef)      1
Pocillopora damicornis                Saudi Arabia  Southern Red Sea            496
Pocillopora verrucosa                 NaN           NaN                          58
Porites compressa                     USA           West Maui                   758
Sinularia flexibilis                  NaN           NaN                           2
Stylophora pistillata                 Saudi Arabia  Southern Red Sea            354
                                      NaN           NaN                         556
Stylophora pistillata (coral)         Taiwan        Green Island                 17
                                                    Ken Ting                     21
                                                    Yeh Liu                       1
Name: id, dtype: int64
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="ncbi-nucleotide-databse">
<h2>NCBI: Nucleotide databse<a class="headerlink" href="#ncbi-nucleotide-databse" title="Permalink to this heading">#</a></h2>
<p>While <code class="docutils literal notranslate"><span class="pre">76</span></code> results is absolute enough to work with, I will expand the scope to include more sequences by searching directly through the nucleotide database.</p>
<p>I will combine the popset dataset and the findings below, and then remove duplicates.</p>
<section id="search-in-ncbi-nucleotide">
<h3>Search in NCBI (Nucleotide)<a class="headerlink" href="#search-in-ncbi-nucleotide" title="Permalink to this heading">#</a></h3>
<p>The review of literature revealed one nucleotide sequence that was not included by the search using <code class="docutils literal notranslate"><span class="pre">search_nucleo(...)</span></code> function, which found a match to a coral species found in the Red Sea <span id="id5">[<a class="reference internal" href="../references.html#id44" title="Claudia Pogoreutz, Clinton A. Oakley, Nils Rädecker, Anny Cárdenas, Gabriela Perna, Nan Xiang, Lifeng Peng, Simon K. Davy, David K. Ngugi, and Christian R. Voolstra. Coral holobiont cues prime endozoicomonas for a symbiotic lifestyle. The ISME Journal, 16(8):1883–1895, Aug 2022. URL: https://doi.org/10.1038/s41396-022-01226-7, doi:10.1038/s41396-022-01226-7.">Pogoreutz <em>et al.</em>, 2022</a>]</span>. We will manually note this, and add it to our list of sequences to download.</p>
<p>Interestingly, looking more into the referenced sequence revealed that the original study had a total of 412 16S Endozoicomonas sequences to work with <span id="id6">[<a class="reference internal" href="../references.html#id47" title="Till Bayer, Matthew J. Neave, Areej Alsheikh-Hussain, Manuel Aranda, Lauren K. Yum, Tracy Mincer, Konrad Hughen, Amy Apprill, and Christian R. Voolstra. The microbiome of the red sea coral stylophora pistillata is dominated by tissue-associated endozoicomonas bacteria. Applied and Environmental Microbiology, 79(15):4759–4762, aug 2013. doi:10.1128/aem.00695-13.">Bayer <em>et al.</em>, 2013</a>]</span>. This will be added by adding the accession numbers by Popset (above).</p>
<p>Additionally, they <span id="id7">[<a class="reference internal" href="../references.html#id47" title="Till Bayer, Matthew J. Neave, Areej Alsheikh-Hussain, Manuel Aranda, Lauren K. Yum, Tracy Mincer, Konrad Hughen, Amy Apprill, and Christian R. Voolstra. The microbiome of the red sea coral stylophora pistillata is dominated by tissue-associated endozoicomonas bacteria. Applied and Environmental Microbiology, 79(15):4759–4762, aug 2013. doi:10.1128/aem.00695-13.">Bayer <em>et al.</em>, 2013</a>]</span> also provided a supplementary phylogenetic tree. We will use this to add more accession ids to our list of sequences to download.</p>
<figure class="align-default" id="parsimony-tree-bayer-2013">
<img alt="../../_images/bayer_2013_fig_s3_1pg_rotated.png" src="../../_images/bayer_2013_fig_s3_1pg_rotated.png" />
<figcaption>
<p><span class="caption-number">Fig. 5 </span><span class="caption-text">Tree showing the phylogenetic relationship of Endozoicomonas to other bacterial species.</span><a class="headerlink" href="#parsimony-tree-bayer-2013" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>These are the accession IDs mentioned in the tree above.</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>Accession ID</p></th>
<th class="head"><p>Status</p></th>
<th class="head"><p>Host</p></th>
<th class="head"><p>DOI</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>GU118644</p></td>
<td><p>Probable</p></td>
<td><p>Montastraea faveolata (coral)</p></td>
<td><p><span id="id8">[<a class="reference internal" href="../references.html#id48" title="Shinichi Sunagawa, Cheryl M. Woodley, and Mónica Medina. Threatened corals provide underexplored microbial habitats. PLoS ONE, 5(3):e9554, mar 2010. doi:10.1371/journal.pone.0009554.">Sunagawa <em>et al.</em>, 2010</a>]</span></p></td>
</tr>
<tr class="row-odd"><td><p>GU118168</p></td>
<td><p>Probable</p></td>
<td><p>Diploria strigosa (coral)</p></td>
<td><p><span id="id9">[<a class="reference internal" href="../references.html#id48" title="Shinichi Sunagawa, Cheryl M. Woodley, and Mónica Medina. Threatened corals provide underexplored microbial habitats. PLoS ONE, 5(3):e9554, mar 2010. doi:10.1371/journal.pone.0009554.">Sunagawa <em>et al.</em>, 2010</a>]</span></p></td>
</tr>
<tr class="row-even"><td><p>GU118379</p></td>
<td><p>Probable</p></td>
<td><p>Gorgonia ventalina (coral)</p></td>
<td><p><span id="id10">[<a class="reference internal" href="../references.html#id48" title="Shinichi Sunagawa, Cheryl M. Woodley, and Mónica Medina. Threatened corals provide underexplored microbial habitats. PLoS ONE, 5(3):e9554, mar 2010. doi:10.1371/journal.pone.0009554.">Sunagawa <em>et al.</em>, 2010</a>]</span></p></td>
</tr>
<tr class="row-odd"><td><p>GU118072</p></td>
<td><p>Probable</p></td>
<td><p>Acropora palmata (coral)</p></td>
<td><p><span id="id11">[<a class="reference internal" href="../references.html#id48" title="Shinichi Sunagawa, Cheryl M. Woodley, and Mónica Medina. Threatened corals provide underexplored microbial habitats. PLoS ONE, 5(3):e9554, mar 2010. doi:10.1371/journal.pone.0009554.">Sunagawa <em>et al.</em>, 2010</a>]</span></p></td>
</tr>
<tr class="row-even"><td><p>GU118404</p></td>
<td><p>Probable</p></td>
<td><p>Gorgonia ventalina (coral)</p></td>
<td><p><span id="id12">[<a class="reference internal" href="../references.html#id48" title="Shinichi Sunagawa, Cheryl M. Woodley, and Mónica Medina. Threatened corals provide underexplored microbial habitats. PLoS ONE, 5(3):e9554, mar 2010. doi:10.1371/journal.pone.0009554.">Sunagawa <em>et al.</em>, 2010</a>]</span></p></td>
</tr>
<tr class="row-odd"><td><p>GU118957</p></td>
<td><p>Probable</p></td>
<td><p>Porites astreoides (coral)</p></td>
<td><p><span id="id13">[<a class="reference internal" href="../references.html#id48" title="Shinichi Sunagawa, Cheryl M. Woodley, and Mónica Medina. Threatened corals provide underexplored microbial habitats. PLoS ONE, 5(3):e9554, mar 2010. doi:10.1371/journal.pone.0009554.">Sunagawa <em>et al.</em>, 2010</a>]</span></p></td>
</tr>
<tr class="row-even"><td><p>GU784983</p></td>
<td><p>Probable</p></td>
<td><p>Sponge Ianthella basta</p></td>
<td><p><span id="id14">[<a class="reference internal" href="../references.html#id57" title="Heidi M. Luter, Steve Whalan, and Nicole S. Webster. Exploring the role of microorganisms in the disease-like syndrome affecting the sponge ianthella basta. Applied and Environmental Microbiology, 76(17):5736–5744, sep 2010. doi:10.1128/aem.00653-10.">Luter <em>et al.</em>, 2010</a>]</span></p></td>
</tr>
<tr class="row-odd"><td><p>GU118966</p></td>
<td><p>Probable</p></td>
<td><p>Porites astreoides (coral)</p></td>
<td><p><span id="id15">[<a class="reference internal" href="../references.html#id48" title="Shinichi Sunagawa, Cheryl M. Woodley, and Mónica Medina. Threatened corals provide underexplored microbial habitats. PLoS ONE, 5(3):e9554, mar 2010. doi:10.1371/journal.pone.0009554.">Sunagawa <em>et al.</em>, 2010</a>]</span></p></td>
</tr>
<tr class="row-even"><td><p>AB695088</p></td>
<td><p>Probable</p></td>
<td><p>Haplosclerida gen. et sp. (purple sponge)</p></td>
<td><p><span id="id16">[<a class="reference internal" href="../references.html#id49" title="Miyuki Nishijima, Kyoko Adachi, Atsuko Katsuta, Yoshikazu Shizuri, and Kazuhide Yamasato. Endozoicomonas numazuensis sp. nov., a gammaproteobacterium isolated from marine sponges, and emended description of the genus endozoicomonas kurahashi and yokota 2007. International Journal of Systematic and Evolutionary Microbiology, 63(Pt_2):709–714, feb 2013. doi:10.1099/ijs.0.042077-0.">Nishijima <em>et al.</em>, 2013</a>]</span></p></td>
</tr>
<tr class="row-odd"><td><p>AM259915</p></td>
<td><p>Probable</p></td>
<td><p>Chondrilla nucula (sponge)</p></td>
<td><p><span id="id17">[<a class="reference internal" href="../references.html#id51" title="Vera Thiel, Sven Leininger, Rolf Schmaljohann, Franz Brümmer, and Johannes F. Imhoff. Sponge-specific bacterial associations of the mediterranean sponge chondrilla nucula (demospongiae, tetractinomorpha). Microbial Ecology, 54(1):101–111, mar 2007. doi:10.1007/s00248-006-9177-y.">Thiel <em>et al.</em>, 2007</a>]</span></p></td>
</tr>
<tr class="row-even"><td><p>DQ884169</p></td>
<td><p>Probable (Uncultured Gammaproteobacteria)</p></td>
<td><p>Cystodytes dellechiajei (tunicate)</p></td>
<td><p><span id="id18">[<a class="reference internal" href="../references.html#id52" title="Manuel Martínez-García, Marta Díaz-Valdés, Gerhard Wanner, Alfonso Ramos-Esplá, and Josefa Antón. Microbial community associated with the colonial ascidian cystodytes dellechiajei. Environmental Microbiology, 9(2):521–534, nov 2006. doi:10.1111/j.1462-2920.2006.01170.x.">Martínez-García <em>et al.</em>, 2006</a>]</span></p></td>
</tr>
<tr class="row-odd"><td><p>DQ884170</p></td>
<td><p>Probable (Gammaproteobacteria)</p></td>
<td><p>Cystodytes dellechiajei (tunicate)</p></td>
<td><p><span id="id19">[<a class="reference internal" href="../references.html#id52" title="Manuel Martínez-García, Marta Díaz-Valdés, Gerhard Wanner, Alfonso Ramos-Esplá, and Josefa Antón. Microbial community associated with the colonial ascidian cystodytes dellechiajei. Environmental Microbiology, 9(2):521–534, nov 2006. doi:10.1111/j.1462-2920.2006.01170.x.">Martínez-García <em>et al.</em>, 2006</a>]</span></p></td>
</tr>
<tr class="row-even"><td><p>DQ884160</p></td>
<td><p>Probable (Uncultured Gammaproteobacteria)</p></td>
<td><p>Cystodytes dellechiajei (tunicate)</p></td>
<td><p><span id="id20">[<a class="reference internal" href="../references.html#id52" title="Manuel Martínez-García, Marta Díaz-Valdés, Gerhard Wanner, Alfonso Ramos-Esplá, and Josefa Antón. Microbial community associated with the colonial ascidian cystodytes dellechiajei. Environmental Microbiology, 9(2):521–534, nov 2006. doi:10.1111/j.1462-2920.2006.01170.x.">Martínez-García <em>et al.</em>, 2006</a>]</span></p></td>
</tr>
<tr class="row-odd"><td><p>DQ917901</p></td>
<td><p>Probable</p></td>
<td><p>Muricea elongata (Octocoral)</p></td>
<td><p><span id="id21">[<a class="reference internal" href="../references.html#id53" title="LK Ranzer, PF Restrepo, and RG Kerr. Data from: microbial community profiles of bleached and wild-type muricea elongata. ncbi popset 134140623. Chemistry and Biochemistry, Florida Atlantic University, 777 Glades Rd., Boca Raton, FL 33431, USA, April 2007. URL: https://www.ncbi.nlm.nih.gov/popset/134140623?report=genbank.">Ranzer <em>et al.</em>, 2007</a>]</span></p></td>
</tr>
<tr class="row-even"><td><p>AY700600</p></td>
<td><p>Probable</p></td>
<td><p>Pocillopora damicornis (coral)</p></td>
<td><p><span id="id22">[<a class="reference internal" href="../references.html#id54" title="David G. Bourne and Colin B. Munn. Diversity of bacteria associated with the coral pocillopora damicornis from the great barrier reef. Environmental Microbiology, 7(8):1162–1174, apr 2005. doi:10.1111/j.1462-2920.2005.00793.x.">Bourne and Munn, 2005</a>]</span></p></td>
</tr>
<tr class="row-odd"><td><p>AY700601</p></td>
<td><p>Probable</p></td>
<td><p>Pocillopora damicornis (coral)</p></td>
<td><p><span id="id23">[<a class="reference internal" href="../references.html#id54" title="David G. Bourne and Colin B. Munn. Diversity of bacteria associated with the coral pocillopora damicornis from the great barrier reef. Environmental Microbiology, 7(8):1162–1174, apr 2005. doi:10.1111/j.1462-2920.2005.00793.x.">Bourne and Munn, 2005</a>]</span></p></td>
</tr>
<tr class="row-even"><td><p>FJ202634</p></td>
<td><p>Probable, (uncultured bacterium)</p></td>
<td><p>Montastraea faveolata (coral)</p></td>
<td><p><span id="id24">[<a class="reference internal" href="../references.html#id59" title="Shinichi Sunagawa, Todd Z DeSantis, Yvette M Piceno, Eoin L Brodie, Michael K DeSalvo, Christian R Voolstra, Ernesto Weil, Gary L Andersen, and Mónica Medina. Bacterial diversity and white plague disease-associated community changes in the caribbean coral montastraea faveolata. The ISME Journal, 3(5):512–521, jan 2009. doi:10.1038/ismej.2008.131.">Sunagawa <em>et al.</em>, 2009</a>]</span></p></td>
</tr>
<tr class="row-odd"><td><p>FJ347758</p></td>
<td><p>Probable</p></td>
<td><p>Montipora aequituberculata (coral)</p></td>
<td><p><span id="id25">[<a class="reference internal" href="../references.html#id60" title="Cho-Song Yang, Ming-Hui Chen, A. B. Arun, Chaolun Allen Chen, Jih-Terng Wang, and Wen-Ming Chen. Endozoicomonas montiporae sp. nov., isolated from the encrusting pore coral montipora aequituberculata. International Journal of Systematic and Evolutionary Microbiology, 60(5):1158–1162, may 2010. doi:10.1099/ijs.0.014357-0.">Yang <em>et al.</em>, 2010</a>]</span></p></td>
</tr>
<tr class="row-even"><td><p>FJ930289</p></td>
<td><p>Probable, (Uncultured bacterium)</p></td>
<td><p>Porites compressa (coral)</p></td>
<td><p><span id="id26">[<a class="reference internal" href="../references.html#id61" title="Mark D. Speck and Stuart P. Donachie. Widespread oceanospirillaceae bacteria in$\less $i$\greater $porites$\less $/i$\greater $spp. Journal of Marine Biology, 2012:1–7, 2012. doi:10.1155/2012/746720.">Speck and Donachie, 2012</a>]</span></p></td>
</tr>
</tbody>
</table>
<blockquote>
<div><p><code class="docutils literal notranslate"><span class="pre">DQ884169</span>&#160; <span class="pre">(C19)</span></code> <code class="docutils literal notranslate"><span class="pre">DQ884170</span> <span class="pre">(C23)</span></code>, <code class="docutils literal notranslate"><span class="pre">DQ884160</span> <span class="pre">(CRNA5)</span></code> noted with “colony sequences C19, C23 and CRNA5 […] formed a group related to Endozoiciomonas elysicola (97.6–99.7% similarity), a bacterium isolated from the marine mollusc Elysia ornata (M. Kurahashi, unpublished) <span id="id27">[<a class="reference internal" href="../references.html#id52" title="Manuel Martínez-García, Marta Díaz-Valdés, Gerhard Wanner, Alfonso Ramos-Esplá, and Josefa Antón. Microbial community associated with the colonial ascidian cystodytes dellechiajei. Environmental Microbiology, 9(2):521–534, nov 2006. doi:10.1111/j.1462-2920.2006.01170.x.">Martínez-García <em>et al.</em>, 2006</a>]</span>”</p>
</div></blockquote>
<p>From above tree, but will not be included.</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>Accession ID</p></th>
<th class="head"><p>Status</p></th>
<th class="head"><p>Host</p></th>
<th class="head"><p>DOI</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>AM503093</p></td>
<td><p>Very unlikely (Marinobacter guineae)</p></td>
<td><p>Antarctic environment</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p>AM229315</p></td>
<td><p>Very unlikely (Halomonas janggokensis)</p></td>
<td><p>saline water</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p>GU291858</p></td>
<td><p>Unlikely (due to distance on tree)</p></td>
<td><p>solar saltern</p></td>
<td><p><span id="id28">[<a class="reference internal" href="../references.html#id58" title="Yochan Joung, Haneul Kim, and Kiseong Joh. Gamma proteobacterium hmd3022 16s ribosomal rna gene, partial sequence. Bioscience and Biotechnology, Hankuk University of Foreign Studies, 89, Wangsan, Mohyun, Cheoin, Yongin, Gyeonggi 449-791, Korea, 2010.">Joung <em>et al.</em>, 2010</a>]</span></p></td>
</tr>
<tr class="row-odd"><td><p>AB205011</p></td>
<td><p>Unlikely, Spongiobacter nickelotolerans</p></td>
<td><p>marine sponge</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p>AB196667</p></td>
<td><p>Probable, not included</p></td>
<td><p>Elysia ornata (sea slug)</p></td>
<td><p><span id="id29">[<a class="reference internal" href="../references.html#id50" title="Midori Kurahashi and Akira Yokota. Endozoicomonas elysicola gen. nov., sp. nov., a γ-proteobacterium isolated from the sea slug elysia ornata. Systematic and Applied Microbiology, 30(3):202–206, apr 2007. doi:10.1016/j.syapm.2006.07.003.">Kurahashi and Yokota, 2007</a>]</span></p></td>
</tr>
<tr class="row-odd"><td><p>DQ917830</p></td>
<td><p>Unlikely (Spongiobacter)</p></td>
<td><p>Cystodytes dellechiajei (tunicate)</p></td>
<td><p><span id="id30">[<a class="reference internal" href="../references.html#id52" title="Manuel Martínez-García, Marta Díaz-Valdés, Gerhard Wanner, Alfonso Ramos-Esplá, and Josefa Antón. Microbial community associated with the colonial ascidian cystodytes dellechiajei. Environmental Microbiology, 9(2):521–534, nov 2006. doi:10.1111/j.1462-2920.2006.01170.x.">Martínez-García <em>et al.</em>, 2006</a>]</span></p></td>
</tr>
<tr class="row-even"><td><p>DQ917877</p></td>
<td><p>Unlikely (Spongiobacter)</p></td>
<td><p>Cystodytes dellechiajei (tunicate)</p></td>
<td><p><span id="id31">[<a class="reference internal" href="../references.html#id52" title="Manuel Martínez-García, Marta Díaz-Valdés, Gerhard Wanner, Alfonso Ramos-Esplá, and Josefa Antón. Microbial community associated with the colonial ascidian cystodytes dellechiajei. Environmental Microbiology, 9(2):521–534, nov 2006. doi:10.1111/j.1462-2920.2006.01170.x.">Martínez-García <em>et al.</em>, 2006</a>]</span></p></td>
</tr>
<tr class="row-odd"><td><p>DQ917879</p></td>
<td><p>Unlikely (Spongiobacter)</p></td>
<td><p>Muricea elongata (Octocoral)</p></td>
<td><p><span id="id32">[<a class="reference internal" href="../references.html#id53" title="LK Ranzer, PF Restrepo, and RG Kerr. Data from: microbial community profiles of bleached and wild-type muricea elongata. ncbi popset 134140623. Chemistry and Biochemistry, Florida Atlantic University, 777 Glades Rd., Boca Raton, FL 33431, USA, April 2007. URL: https://www.ncbi.nlm.nih.gov/popset/134140623?report=genbank.">Ranzer <em>et al.</em>, 2007</a>]</span></p></td>
</tr>
<tr class="row-even"><td><p>GQ853555</p></td>
<td><p>Probable, not included</p></td>
<td><p>Loripes lacteus (clam)</p></td>
<td><p>gill symbiont</p></td>
</tr>
<tr class="row-odd"><td><p>FM162182</p></td>
<td><p>Probable, not included</p></td>
<td><p>Bathymodiolus brooksi (mussels)</p></td>
<td><p><span id="id33">[<a class="reference internal" href="../references.html#id56" title="Frank U. Zielinski, Annelie Pernthaler, Sébastien Duperron, Luciana Raggi, Olav Giere, Christian Borowski, and Nicole Dubilier. Widespread occurrence of an intranuclear bacterial parasite in vent and seep bathymodiolin mussels. Environmental Microbiology, 11(5):1150–1167, apr 2009. doi:10.1111/j.1462-2920.2008.01847.x.">Zielinski <em>et al.</em>, 2009</a>]</span></p></td>
</tr>
<tr class="row-even"><td><p>FM163188</p></td>
<td><p>Probable, not included</p></td>
<td><p>Bathymodiolus brooksi (mussels)</p></td>
<td><p><span id="id34">[<a class="reference internal" href="../references.html#id56" title="Frank U. Zielinski, Annelie Pernthaler, Sébastien Duperron, Luciana Raggi, Olav Giere, Christian Borowski, and Nicole Dubilier. Widespread occurrence of an intranuclear bacterial parasite in vent and seep bathymodiolin mussels. Environmental Microbiology, 11(5):1150–1167, apr 2009. doi:10.1111/j.1462-2920.2008.01847.x.">Zielinski <em>et al.</em>, 2009</a>]</span></p></td>
</tr>
<tr class="row-odd"><td><p>FJ154998</p></td>
<td><p>Probable, (uncultured bacterium) not included</p></td>
<td><p>ocean water</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p>EU884930</p></td>
<td><p>Probable, not included</p></td>
<td><p>Sixbar angelfish</p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<p>Here are other accession IDs that were found when examining the references of the papers above.</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>Accession ID</p></th>
<th class="head"><p>DOI</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>AB695089</p></td>
<td><p><span id="id35">[<a class="reference internal" href="../references.html#id49" title="Miyuki Nishijima, Kyoko Adachi, Atsuko Katsuta, Yoshikazu Shizuri, and Kazuhide Yamasato. Endozoicomonas numazuensis sp. nov., a gammaproteobacterium isolated from marine sponges, and emended description of the genus endozoicomonas kurahashi and yokota 2007. International Journal of Systematic and Evolutionary Microbiology, 63(Pt_2):709–714, feb 2013. doi:10.1099/ijs.0.042077-0.">Nishijima <em>et al.</em>, 2013</a>]</span></p></td>
</tr>
</tbody>
</table>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">search_nucleo</span><span class="p">(</span><span class="n">limit</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">additional_ids</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]):</span>
<span class="w">  </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">  Search for the nucleotide database for nucleotides that mention coral and Endozoicomonas.</span>

<span class="sd">  :param `limit`: The maximum number of results to return.</span>
<span class="sd">  &quot;&quot;&quot;</span>

  <span class="n">term</span> <span class="o">=</span> <span class="s2">&quot;(coral[All Fields] AND 16S[Title]) AND Endozoicomonas[Organism]&quot;</span>
  <span class="k">try</span><span class="p">:</span>
    <span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span>
      <span class="n">term</span><span class="o">=</span><span class="n">term</span><span class="p">,</span>
      <span class="n">retmax</span><span class="o">=</span><span class="n">limit</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">DictionaryElement</span><span class="p">):</span>
      <span class="n">ids</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="s2">&quot;IdList&quot;</span><span class="p">]</span>
      <span class="n">ids</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">additional_ids</span><span class="p">)</span>
      <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">ids</span><span class="p">)</span>
      <span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">ids</span><span class="p">,</span> <span class="n">rettype</span><span class="o">=</span><span class="s2">&quot;gb&quot;</span><span class="p">)</span>
      <span class="n">records</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="s2">&quot;gb&quot;</span><span class="p">))</span>
      <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="n">records</span><span class="p">),</span> <span class="n">ids</span><span class="p">,</span> <span class="n">records</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="k">return</span> <span class="mi">0</span><span class="p">,</span> <span class="p">[],</span> <span class="p">[]</span>
  <span class="k">except</span> <span class="n">HTTPError</span> <span class="k">as</span> <span class="n">http_e</span><span class="p">:</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">http_e</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">exception</span><span class="p">(</span><span class="n">http_e</span><span class="p">)</span>
    <span class="k">return</span> <span class="mi">0</span><span class="p">,</span> <span class="p">[],</span> <span class="p">[]</span>
  <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">exception</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
    <span class="k">return</span> <span class="mi">0</span><span class="p">,</span> <span class="p">[],</span> <span class="p">[]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">additional_accessions</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AB695089&#39;</span><span class="p">]</span>
<span class="n">count_nuc</span><span class="p">,</span> <span class="n">idlist_nuc</span><span class="p">,</span> <span class="n">records_nuc</span> <span class="o">=</span> <span class="n">search_nucleo</span><span class="p">(</span><span class="mi">4000</span><span class="p">,</span> <span class="n">additional_accessions</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="summary-of-nucleotide-database">
<h3>Summary of nucleotide database<a class="headerlink" href="#summary-of-nucleotide-database" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pandas</span> <span class="kn">import</span> <span class="n">concat</span>

<span class="n">DEBUG</span> <span class="o">=</span> <span class="kc">False</span>
<span class="n">feat_df_nuc0</span><span class="p">,</span> <span class="n">recs_nuc</span> <span class="o">=</span> <span class="n">table_features</span><span class="p">(</span><span class="n">count_nuc</span><span class="p">,</span> <span class="n">records_nuc</span><span class="p">,</span> <span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="n">DEBUG</span><span class="p">)</span>
<span class="n">feat_df_nuc0</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dir_path</span><span class="p">,</span> <span class="s2">&quot;features_of_nucleotide_seq.csv&quot;</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO:download_16s:Records to save: 154/154
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">feat_df_all0</span> <span class="o">=</span> <span class="n">concat</span><span class="p">([</span><span class="n">feat_df_nuc0</span><span class="p">,</span> <span class="n">feat_df_pop0</span><span class="p">])</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="s2">&quot;id&quot;</span><span class="p">)</span>
<span class="n">feat_df_all0</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dir_path</span><span class="p">,</span> <span class="s2">&quot;feat_all_seq.csv&quot;</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">unique</span><span class="p">(</span><span class="n">to_combine</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">]],</span> <span class="n">all_df</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">):</span>
    <span class="n">idlist</span> <span class="o">=</span> <span class="n">all_df</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
    <span class="c1"># intilize a null list</span>
    <span class="n">keep_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="n">todo_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="n">already_found_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">todo</span> <span class="ow">in</span> <span class="n">to_combine</span><span class="p">:</span>
        <span class="n">todo_list</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">todo</span><span class="p">)</span>

    <span class="c1"># traverse for all elements</span>
    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">todo_list</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">already_found_list</span><span class="p">:</span>
          <span class="k">continue</span>
        <span class="k">else</span><span class="p">:</span>
          <span class="n">already_found_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
        <span class="c1"># check if exists in unique_list or not</span>
        <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">idlist</span><span class="p">:</span>
            <span class="n">keep_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">keep_list</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">unique_recs</span> <span class="o">=</span> <span class="n">unique</span><span class="p">([</span><span class="n">recs_nuc</span><span class="p">,</span> <span class="n">recs_pop</span><span class="p">],</span> <span class="n">feat_df_all0</span><span class="p">)</span>

<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unique records found: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">unique_recs</span><span class="p">)</span><span class="si">}</span><span class="s2"> / </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">feat_df_all0</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">())</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="c1"># Save the records to a file</span>
<span class="n">save_file</span><span class="p">(</span><span class="s2">&quot;all_nucleotide.gb&quot;</span><span class="p">,</span> <span class="n">unique_recs</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">)</span>
<span class="n">save_file</span><span class="p">(</span><span class="s2">&quot;all_nucleotide.fasta&quot;</span><span class="p">,</span> <span class="n">unique_recs</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO:download_16s:Unique records found: 1854 / 1854
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="n">DEBUG</span><span class="p">:</span>
  <span class="n">col_interest</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">,</span> <span class="s1">&#39;organism&#39;</span><span class="p">,</span> <span class="s1">&#39;strain&#39;</span><span class="p">,</span> <span class="s1">&#39;host&#39;</span><span class="p">,</span> <span class="s1">&#39;country&#39;</span><span class="p">,</span> <span class="s1">&#39;pubmed_ids&#39;</span><span class="p">,</span> <span class="s1">&#39;product&#39;</span><span class="p">,</span> <span class="s1">&#39;isolation_source&#39;</span><span class="p">]</span>
<span class="k">else</span><span class="p">:</span>
  <span class="n">col_interest</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">,</span> <span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="s1">&#39;organism&#39;</span><span class="p">,</span> <span class="s1">&#39;host&#39;</span><span class="p">,</span> <span class="s1">&#39;country&#39;</span><span class="p">,</span> <span class="s1">&#39;pubmed_ids&#39;</span><span class="p">,</span> <span class="s1">&#39;isolation_source&#39;</span><span class="p">,</span> <span class="s1">&#39;iso_source2&#39;</span><span class="p">,</span> <span class="s1">&#39;product&#39;</span><span class="p">]</span>

<span class="n">feat_df_all</span> <span class="o">=</span> <span class="n">feat_df_all0</span><span class="p">[</span><span class="n">col_interest</span><span class="p">]</span>
<span class="n">feat_df_all</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>id</th>
      <th>group</th>
      <th>organism</th>
      <th>host</th>
      <th>country</th>
      <th>pubmed_ids</th>
      <th>isolation_source</th>
      <th>iso_source2</th>
      <th>product</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>OM273416.1</td>
      <td>OM273</td>
      <td>Endozoicomonas euniceicola</td>
      <td>Eunicea flexuosa</td>
      <td>USA</td>
      <td></td>
      <td>Florida Keys</td>
      <td>Florida Keys</td>
      <td>16S ribosomal RNA</td>
    </tr>
    <tr>
      <th>1</th>
      <td>OM273412.1</td>
      <td>OM273</td>
      <td>Endozoicomonas gorgoniicola</td>
      <td>Eunicea flexuosa</td>
      <td>USA</td>
      <td></td>
      <td>Florida Keys</td>
      <td>Florida Keys</td>
      <td>16S ribosomal RNA</td>
    </tr>
    <tr>
      <th>2</th>
      <td>OL957540.1</td>
      <td>OL957</td>
      <td>uncultured Endozoicomonas sp.</td>
      <td>Stylophora pistillata</td>
      <td>Japan</td>
      <td></td>
      <td>Okinawa</td>
      <td>Okinawa</td>
      <td>16S ribosomal RNA</td>
    </tr>
    <tr>
      <th>3</th>
      <td>OL957539.1</td>
      <td>OL957</td>
      <td>uncultured Endozoicomonas sp.</td>
      <td>Stylophora pistillata</td>
      <td>Japan</td>
      <td></td>
      <td>Okinawa</td>
      <td>Okinawa</td>
      <td>16S ribosomal RNA</td>
    </tr>
    <tr>
      <th>4</th>
      <td>OL957538.1</td>
      <td>OL957</td>
      <td>uncultured Endozoicomonas sp.</td>
      <td>Stylophora pistillata</td>
      <td>Japan</td>
      <td></td>
      <td>Okinawa</td>
      <td>Okinawa</td>
      <td>16S ribosomal RNA</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="id36">
<h3>Geographic details<a class="headerlink" href="#id36" title="Permalink to this heading">#</a></h3>
<p>Here shows the breakdown by country, and by isolation source as reported by the metadata.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">feat_df_all</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;country&#39;</span><span class="p">],</span> <span class="n">dropna</span><span class="o">=</span><span class="kc">False</span><span class="p">)[</span><span class="s1">&#39;country&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>country
Bahamas           1
France            1
Japan            40
Malaysia          5
Saudi Arabia    586
Taiwan           69
USA             786
NaN               0
Name: country, dtype: int64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">feat_df_all</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;country&#39;</span><span class="p">,</span> <span class="s1">&#39;isolation_source&#39;</span><span class="p">],</span> <span class="n">dropna</span><span class="o">=</span><span class="kc">False</span><span class="p">)[</span><span class="s1">&#39;isolation_source&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>country       isolation_source        
Bahamas       Tuna Alley (inside reef)      1
France        NaN                           0
Japan         Okinawa                      39
              Shizuoka, Numazu              1
Malaysia      NaN                           0
Saudi Arabia  Southern Red Sea            586
Taiwan        Green Island                 17
              Ken Ting                     21
              Kenting                      28
              Yeh Liu                       1
              NaN                           0
USA           Florida Keys                  2
              Hillsboro Ledge               1
              Summerland Key, Florida      25
              West Maui                   758
NaN           NaN                           0
Name: isolation_source, dtype: int64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">feat_df_all</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;country&#39;</span><span class="p">,</span> <span class="s1">&#39;isolation_source&#39;</span><span class="p">,</span> <span class="s1">&#39;host&#39;</span><span class="p">],</span> <span class="n">dropna</span><span class="o">=</span><span class="kc">False</span><span class="p">)[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>country       isolation_source          host                                                   
Bahamas       Tuna Alley (inside reef)  Plexaura sp. (gorgonian coral)                               1
France        NaN                       Corallium rubrum                                             1
Japan         Okinawa                   Stylophora pistillata                                       39
              Shizuoka, Numazu          Isolated from purple sponge (Haplosclerida gen. et sp.)      1
Malaysia      NaN                       Marine tunicates                                             5
Saudi Arabia  Southern Red Sea          Acropora humilis                                           161
                                        Pocillopora damicornis                                     248
                                        Stylophora pistillata                                      177
Taiwan        Green Island              Stylophora pistillata (coral)                               17
              Ken Ting                  Stylophora pistillata (coral)                               21
              Kenting                   Stylophora pistillata                                       28
              Yeh Liu                   Stylophora pistillata (coral)                                1
              NaN                       sea coral                                                    2
USA           Florida Keys              Eunicea flexuosa                                             2
              Hillsboro Ledge           Eunicea fusca (gorgonian coral)                              1
              Summerland Key, Florida   Plexaura homomalla (gorgonian coral)                        25
              West Maui                 Porites compressa                                          758
NaN           NaN                       Acropora clathrata                                           1
                                        Acropora cytherea                                            1
                                        Acropora hemprichii                                          1
                                        Eunicella cavolini                                          20
                                        Lobophytum compactum                                         1
                                        Pocillopora verrucosa                                       58
                                        Sea coral                                                    1
                                        Sinularia flexibilis                                         2
                                        Stylophora pistillata                                      278
                                        sea coral                                                    3
Name: id, dtype: int64
</pre></div>
</div>
</div>
</div>
</section>
<section id="fixing-isolation-source-column">
<h3>Fixing isolation_source column<a class="headerlink" href="#fixing-isolation-source-column" title="Permalink to this heading">#</a></h3>
<p>By the inspection above, we know that we need make <a class="reference download internal" download="" href="../../_downloads/b306d834f9bf70f1df824af815e9e870/fix_list_gran_5.csv"><span class="xref download myst">a list</span></a> of <code class="docutils literal notranslate"><span class="pre">isolation_source</span></code> rows to fix manually by reading the original literature it was published in.</p>
<p><code class="docutils literal notranslate"><span class="pre">OL957</span></code> based results can be traced back to this paper by <span id="id37">[<a class="reference internal" href="../references.html#id43" title="Naohisa Wada, Ming-Tsung Hsu, Kshitij Tandon, Silver Sung-Yun Hsiao, Hsing-Ju Chen, Yu-Hsiang Chen, Pei-Wen Chiang, Sheng-Ping Yu, Chih-Ying Lu, Yu-Jing Chiou, Yung-Chi Tu, Xuejiao Tian, Bi-Chang Chen, Der-Chuen Lee, Hideyuki Yamashiro, David G. Bourne, and Sen-Lin Tang. High-resolution spatial and genomic characterization of coral-associated microbial aggregates in the coral stylophora pistillata. Science Advances, jul 2022. doi:10.1126/sciadv.abo2431.">Wada <em>et al.</em>, 2022</a>]</span>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">create_fix_csv</span><span class="p">(</span><span class="n">df</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">,</span> <span class="n">csv_file_name</span><span class="o">=</span><span class="s2">&quot;fix_list.csv&quot;</span><span class="p">):</span>
  <span class="n">manual_fix</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;country&#39;</span><span class="p">,</span> <span class="s1">&#39;isolation_source&#39;</span><span class="p">,</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="s1">&#39;id&#39;</span><span class="p">]</span>
  <span class="n">df2</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">manual_fix</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
  <span class="n">df2</span> <span class="o">=</span> <span class="n">df2</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;country&#39;</span><span class="p">,</span> <span class="s1">&#39;isolation_source&#39;</span><span class="p">,</span><span class="s1">&#39;group&#39;</span><span class="p">],</span> <span class="n">dropna</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">first</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

  <span class="n">df3</span> <span class="o">=</span> <span class="n">merge</span><span class="p">(</span><span class="n">df2</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">)</span>
  <span class="n">df3</span><span class="p">[</span><span class="n">manual_fix</span><span class="p">]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dir_path</span><span class="p">,</span> <span class="n">csv_file_name</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">index_label</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">na_rep</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

  <span class="k">return</span> <span class="n">df3</span>

<span class="n">feat_df_nuc2</span> <span class="o">=</span> <span class="n">create_fix_csv</span><span class="p">(</span><span class="n">feat_df_all</span><span class="p">,</span> <span class="s2">&quot;fix_list_gran_5.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./src\4_full_analysis"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  <!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="../3_pilot_3/tree_finding.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Tree finding</p>
      </div>
    </a>
    <a class="right-next"
       href="download_genome.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Download genomes</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#setup-the-environment">Setup the environment</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#clean-and-wrangle-tables">Clean and wrangle tables</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#ncbi-popset">NCBI: Popset</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#search-in-ncbi-popset">Search in NCBI (Popset)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#summary-of-popset">Summary of popset</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#geographic-details">Geographic details</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#ncbi-nucleotide-databse">NCBI: Nucleotide databse</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#search-in-ncbi-nucleotide">Search in NCBI (Nucleotide)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#summary-of-nucleotide-database">Summary of nucleotide database</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id36">Geographic details</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#fixing-isolation-source-column">Fixing isolation_source column</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Kaede Ito
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=ac02cc09edc035673794"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=ac02cc09edc035673794"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>