{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Align and Download"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We are using the package biopython to download nucleotide datasets from NCBI's Genbank, then prep those to align later."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The guide on downloading from NCBI is from [this guide](https://notebook.community/widdowquinn/Notebooks-Bioinformatics/Biopython_NCBI_Entrez_downloads)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "# from Bio.Align.Applications import ClustalwCommandline\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a new directory (if needed) for output/downloads\n",
    "os.makedirs(\"ncbi_downloads\", exist_ok=True)\n",
    "\n",
    "out_amount = os.listdir(\"ncbi_downloads\")\n",
    "initiate_download = False\n",
    "\n",
    "if len(out_amount) == 0:\n",
    "    raise \"No organism list in ncbi_downloads.\"\n",
    "\n",
    "if len(out_amount) == 1:\n",
    "    initiate_download = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "Entrez.email = \"kaedeito@student.ubc.ca\"\n",
    "# This line sets the name of the tool that is making the queries\n",
    "Entrez.tool = \"alignment.ipynb\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sample search to download from NCBI\n",
    "Here is an example using the generic search term of \"Endozoicomonas\" and \"coral\" in NCBI's Genbank nucleotide database.\n",
    "\n",
    "The `Entrez.esearch(...)` function is documented [here](https://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ESearch)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['NZ_CP120717.1', 'CP120717.1', 'NZ_JAHHPM000000000.1', 'NZ_JAHHPM010000603.1', 'NZ_JAHHPM010000602.1', 'NZ_JAHHPM010000601.1', 'NZ_JAHHPM010000600.1', 'NZ_JAHHPM010000599.1', 'NZ_JAHHPM010000598.1', 'NZ_JAHHPM010000597.1']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "term = \"Endozoicomonas[Organism] AND coral[All Fields] AND biomol_genomic[PROP]\"\n",
    "handle = Entrez.esearch(db=\"nucleotide\",\n",
    "  term=term,\n",
    "  retmax=10,\n",
    "  idtype=\"acc\",\n",
    ")\n",
    "\n",
    "# This line converts the returned information from NCBI into a form we\n",
    "# can use, as before.\n",
    "record = Entrez.read(handle)\n",
    "print(record['IdList'])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download sequences for our organism list\n",
    "From a file called \"organism_list.txt\", read the list of organisms to search with.\n",
    "Add the search term \"coral\" to the end of each organism name, and limit to 10 results per taxon."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "list_orgs = open(\"ncbi_downloads/organism_list.txt\").read().splitlines()\n",
    "\n",
    "list_of_accession_id = []\n",
    "\n",
    "min_seq_length = 1200\n",
    "max_seq_length = 10000\n",
    "\n",
    "length_search = f\"({min_seq_length}[SLEN] : {max_seq_length}[SLEN])\"\n",
    "\n",
    "\n",
    "if initiate_download:\n",
    "  for org in list_orgs:\n",
    "    term = f\"(\\\"{org}\\\"[Organism] OR {org}[All Fields]) AND coral[All Fields] AND {length_search} AND biomol_genomic[PROP]\"\n",
    "    handle = Entrez.esearch(db=\"nucleotide\",\n",
    "      term=term,\n",
    "      retmax=10,\n",
    "      idtype=\"acc\",\n",
    "    )\n",
    "    record = Entrez.read(handle)\n",
    "    list_of_accession_id.append(record['IdList'])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A class to hold the search results, and a function to print the results to a fasta file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dataclasses import dataclass\n",
    "\n",
    "@dataclass\n",
    "class OrganismAccession:\n",
    "  ids: list[str]\n",
    "  organism: str\n",
    "  handle: str\n",
    "  folder: str\n",
    "\n",
    "  def __init__(self, ids, organism):\n",
    "    self.ids = ids\n",
    "    self.organism = organism\n",
    "    self.folder = f\"ncbi_downloads\\\\{self.organism}\"\n",
    "    os.makedirs(self.folder, exist_ok=True)\n",
    "    self.handle = Entrez.efetch(db=\"nucleotide\", id=ids, rettype=\"fasta\", retmode=\"text\")\n",
    "\n",
    "  def download_save(self):\n",
    "    with open(f\"{self.folder}\\\\nucleotide.fasta\", \"w\") as out_handle:\n",
    "      out_handle.write(self.handle.read())\n",
    "      print(f\"all\\\\{self.organism} Saved\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all\\Endozoicomonas sp. ONNA2 Saved\n",
      "all\\Endozoicomonas acroporae Saved\n",
      "all\\Endozoicomonas sp. ONNA1 Saved\n",
      "all\\Endozoicomonas sp. SESOKO3 Saved\n",
      "all\\Endozoicomonas sp. SESOKO2 Saved\n",
      "all\\Endozoicomonas sp. SESOKO4 Saved\n",
      "all\\Endozoicomonas sp. YOMI1 Saved\n",
      "all\\Endozoicomonas sp. SESOKO1 Saved\n",
      "all\\Endozoicomonas sp. ISHI1 Saved\n",
      "all\\Endozoicomonas montiporae Saved\n",
      "all\\Endozoicomonas sp. G2_1 Saved\n",
      "all\\Endozoicomonas sp. Saved\n",
      "all\\Endozoicomonas sp. G2_2 Saved\n",
      "all\\Endozoicomonas coralli Saved\n",
      "all\\Endozoicomonas gorgoniicola Saved\n",
      "all\\Endozoicomonas euniceicola Saved\n"
     ]
    }
   ],
   "source": [
    "if initiate_download:\n",
    "  len_org = len(list_orgs)\n",
    "  len_results = len(list_of_accession_id)\n",
    "  if len_org != len_results:\n",
    "    raise Exception(\"Length of organism list and results list are not the same\")\n",
    "\n",
    "  for i in range(len_org):\n",
    "    org = list_orgs[i]\n",
    "    ids = list_of_accession_id[i]\n",
    "    organism_accession = OrganismAccession(ids, org)\n",
    "    organism_accession.download_save()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
