

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Proposal &#8212; Coral Bleaching - BIOL 417</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=ac02cc09edc035673794" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=ac02cc09edc035673794" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=ac02cc09edc035673794" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=ac02cc09edc035673794" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=ac02cc09edc035673794" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=ac02cc09edc035673794" />
  <script src="../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=ac02cc09edc035673794"></script>

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'src/proposal';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="List of used datasets for analysis" href="exploratory/genome/index.html" />
    <link rel="prev" title="Coral microbiome dynamics, functions and design in a changing world" href="exploratory/van_Oppen_2019.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../readme.html">
  
  
  
  
  
    <p class="title logo__title">Coral Bleaching - BIOL 417</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../readme.html">
                    Welcome to BIOL 417 101 Capstone!
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Exploring Literature</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="exploratory/index.html">Proposal ideas</a></li>
<li class="toctree-l1"><a class="reference internal" href="exploratory/Pollock_2018.html">Coral-associated bacteria demonstrate phylosymbiosis and cophylogeny</a></li>
<li class="toctree-l1"><a class="reference internal" href="exploratory/Ziegler_2019.html">Coral bacterial community structure responds to environmental change in a host-specific manner</a></li>
<li class="toctree-l1"><a class="reference internal" href="exploratory/Leite_2018.html">Specific plasmid patterns and high rates of bacterial co-occurrence within the coral holobiont</a></li>
<li class="toctree-l1"><a class="reference internal" href="exploratory/Swain_2021.html">A Phylogeny-Informed Analysis of the Global Coral-Symbiodiniaceae Interaction Network Reveals that Traits Correlated with Thermal Bleaching Are Specific to Symbiont Transmission Mode</a></li>
<li class="toctree-l1"><a class="reference internal" href="exploratory/van_Oppen_2019.html">Coral microbiome dynamics, functions and design in a changing world</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Proposal</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Proposal</a></li>
<li class="toctree-l1"><a class="reference internal" href="exploratory/genome/index.html">List of used datasets for analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="alignment.html">Align and Download</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">End</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/src/proposal.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Proposal</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#id4"></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#planned-procedure">Planned procedure</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#for-plasmid">For plasmid:</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#for-regular-tree">For regular tree:</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#comparison">Comparison:</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="proposal">
<h1>Proposal<a class="headerlink" href="#proposal" title="Permalink to this heading">#</a></h1>
<p>The species of interest was of coral. Through the initial dive of literature of current coral research, the topic of the effects of the coral microbiome and its implications on the health of it was intriguing and led to more digging on this topic.</p>
<p>As the author of <span id="id1">Pollock <em>et al.</em> [<a class="reference internal" href="references.html#id8" title="F. Joseph Pollock, Ryan McMinds, Styles Smith, David G. Bourne, Bette L. Willis, Mónica Medina, Rebecca Vega Thurber, and Jesse R. Zaneveld. Coral-associated bacteria demonstrate phylosymbiosis and cophylogeny. Nature Communications, nov 2018. doi:10.1038/s41467-018-07275-x.">2018</a>]</span> states, the coral’s mucus, tissue, and skeleton show distinct microbial community composition. In combination with a pattern called phylosymbiosis (correlation found between the host’s own phylogenetics relationship and those of their own microbiome) they created a strategy to analyze the Australian coral’s distinct microbial communities within each major anatomical compartments respective to the host phylogeny, host functional traits, geography and environmental variables. What they find is striking: the relative influence of host vs. environmental parameters varies strongly across anatomy. For example, for mucus, microbiome was relatively stable across colonies, and boasted higher abundance of core microbes vs variable ones. In addition, the mucus’ microbiome was more strongly affected by environmental factors (e.g., season, temperature, and turf algal competition) than tissue or skeleton microbiomes. To further cement this, the latitude of the sampling location significantly influenced the richness and composition for coral mucus and tissue, while not so strong for skeleton. This agrees with the other finding around environmental factors, as these factors vary wildly across latitude, which could be the link between the results. Finally, the signal for phylosymbiosis was only significant in the skeleton. This agrees with the other finding that host specificity in the mucus microbiome was limited to relatively recent divergences, and the patterns does not hold to a significant extent across larger scales of phylogeny of the coral (host).</p>
<p>A different study by <span id="id2">Ziegler <em>et al.</em> [<a class="reference internal" href="references.html#id9" title="Maren Ziegler, Carsten G. B. Grupstra, Marcelle M. Barreto, Martin Eaton, Jaafar BaOmar, Khalid Zubier, Abdulmohsin Al-Sofyani, Adnan J. Turki, Rupert Ormond, and Christian R. Voolstra. Coral bacterial community structure responds to environmental change in a host-specific manner. Nature Communications, jul 2019. doi:10.1038/s41467-019-10969-5.">2019</a>]</span> looked at how the bacterial communities respond in changes to the environment. They note studying this is important because these microbiomes fulfill a range of functions including nitrogen fixation, sulfur cycling, and protection against pathogenic bacteria. They can also help with abundance in certain species of coral just with the presence/absence of specific bacteria. Their work examined whether coral microbiomes could recover from chronic pollution and focused on two species’ microbiome (A. hemprichii, P. verrucosa). These two species responded fairly differently, one species was highly flexible and more variable (a generalist), and one that was fairly stable and overall, less variable in response to changing environmental conditions (a specialist). They concluded that coral microbiome can be broadly classified into two major categories. One, where they labeled as “microbiome conformer” (showing microbial adaptation to the surrounding environment) where they noted that the drawbacks were relatively slow growth rate. The second where the coral was a “microbiome regulator” (showing microbial regulation that maintains a constant microbiome) but the coral was limited in physiological plasticity and ecological niche space but afforded them faster growth rate.</p>
<p>Another study that examined coral microbiome by <span id="id3">Leite <em>et al.</em> [<a class="reference internal" href="references.html#id11" title="Deborah C. A. Leite, Joana F. Salles, Emiliano N. Calderon, Jan D. van Elsas, and Raquel S. Peixoto. Specific plasmid patterns and high rates of bacterial co-occurrence within the coral holobiont. Ecology and Evolution, 8(3):1818–1832, jan 2018. doi:10.1002/ece3.3717.">2018</a>]</span> looked at the co‐occurrence patterns in microbiomes which could help with estimating of alpha and beta diversity. They specifically surveyed two broad‐host‐range (BHR) plasmid groups (IncP‐1B and PromA) as they’ve been used to evaluate the horizontal transfer of genetic material across multiple different environments and different hosts. Their results were that the holobiont selects the IncP‐1 and PromA (BHR groups of plasmids) only in coral mucus. Additionally, based on co‐occurrence among Proteobacteria and Firmicutes members they found many related operational taxonomic units (OTUs). The posit that the explanation could be that:</p>
<p>a) Something about these specific taxa is so important they cannot be replaced by other taxa / horizontally transferred genes (strong dependency)</p>
<p>b) Alternatively, or complementarily, the interactions within the community (which may include the host) maintain their presence / absence (symbiotic/competitive etc. relationships)</p>
<p>c) the vertical transmission of microbiome is largely what explains the correlation (legacy imposed by ancestors is what drives composition)</p>
<section id="id4">
<h2><a class="headerlink" href="#id4" title="Permalink to this heading">#</a></h2>
<blockquote>
<div><p>conformer/regulator X plasmid phylogenetic tree/regular tree</p>
</div></blockquote>
<p>Can the plasmid’s phylogenetic tree be a better measure of the coral’s microbiome level of “core” bacteria vs “variable” bacteria than the regular phylogenetic tree?
Use plasmid’s phylogenetic tree as a proxy measure for the level of “conformer” / “regulator” a coral’s microbiome is?</p>
<table class="table">
<thead>
<tr class="row-odd"><th class="head text-right"><p></p></th>
<th class="head"><p>conformer</p></th>
<th class="head"><p>regulator</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-right"><p>plasmid phylogenetic tree</p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td class="text-right"><p>regular tree</p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<ul class="simple">
<li><p><a class="reference external" href="https://link.springer.com/article/10.1186/s40168-023-01653-4">https://link.springer.com/article/10.1186/s40168-023-01653-4</a></p></li>
<li><p><a class="reference external" href="https://www.sciencedirect.com/science/article/pii/S0048969723064306">https://www.sciencedirect.com/science/article/pii/S0048969723064306</a></p></li>
<li><p><a class="reference external" href="https://www.nature.com/articles/s41396-022-01283-y">https://www.nature.com/articles/s41396-022-01283-y</a></p></li>
<li><p><a class="reference external" href="https://link.springer.com/chapter/10.1007/978-3-030-97189-2_3">https://link.springer.com/chapter/10.1007/978-3-030-97189-2_3</a></p></li>
<li><p><a class="reference external" href="https://www.sciencedirect.com/science/article/pii/S0048969723028796">https://www.sciencedirect.com/science/article/pii/S0048969723028796</a></p></li>
<li><p><a class="reference external" href="https://jms.mabjournal.com/index.php/mab/article/view/2162">https://jms.mabjournal.com/index.php/mab/article/view/2162</a></p></li>
<li><p><a class="reference external" href="http://journal.article2publish.com/id/eprint/2403/">http://journal.article2publish.com/id/eprint/2403/</a></p></li>
<li><p><a class="reference external" href="https://www.biorxiv.org/content/10.1101/2023.01.31.526474v1.full">https://www.biorxiv.org/content/10.1101/2023.01.31.526474v1.full</a></p></li>
<li><p><a class="reference external" href="https://www.pnas.org/doi/10.1073/pnas.1920779117#sec-2">https://www.pnas.org/doi/10.1073/pnas.1920779117#sec-2</a></p></li>
<li><p><a class="reference external" href="https://journals.asm.org/doi/10.1128/msystems.00921-22">https://journals.asm.org/doi/10.1128/msystems.00921-22</a></p></li>
<li><p><a class="reference external" href="https://www.nature.com/articles/s42003-023-04590-y">https://www.nature.com/articles/s42003-023-04590-y</a></p></li>
<li><p><a class="reference external" href="https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-023-01622-x#Sec13">https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-023-01622-x#Sec13</a></p></li>
<li><p><a class="reference external" href="https://www.biorxiv.org/content/10.1101/334409v2.full">https://www.biorxiv.org/content/10.1101/334409v2.full</a></p></li>
<li><p><a class="reference external" href="https://www.frontiersin.org/articles/10.3389/fmicb.2018.02167/full">https://www.frontiersin.org/articles/10.3389/fmicb.2018.02167/full</a></p></li>
<li><p><a class="reference external" href="https://academic.oup.com/nar/article/51/15/e83/7222081">https://academic.oup.com/nar/article/51/15/e83/7222081</a></p></li>
<li><p><a class="reference external" href="https://www.ncbi.nlm.nih.gov/biosample?Db=biosample&amp;amp;DbFrom=bioproject&amp;amp;Cmd=Link&amp;amp;LinkName=bioproject_biosample&amp;amp;LinkReadableName=BioSample&amp;amp;ordinalpos=1&amp;amp;IdsFromResult=495243">https://www.ncbi.nlm.nih.gov/biosample?Db=biosample&amp;DbFrom=bioproject&amp;Cmd=Link&amp;LinkName=bioproject_biosample&amp;LinkReadableName=BioSample&amp;ordinalpos=1&amp;IdsFromResult=495243</a></p></li>
</ul>
</section>
<section id="planned-procedure">
<h2>Planned procedure<a class="headerlink" href="#planned-procedure" title="Permalink to this heading">#</a></h2>
<p>While keeping the anatomy variability in mind, I want to compare the phylogenetic tree made from tracing plasmids (plasmid phylogenetic tree) vs a regular phylogenetic tree of the microbiome found between species that could be classified as a “microbiome conformer” vs a “microbiome regulator”. (If further narrowing down is needed, maybe focus on anatomy?). For this, it would make sense to refer to the raw sequencing files used by Pollock et al. (2018) (accessible under European Nucleotide Archive with the accession number PRJEB28183). This will require:</p>
<ol class="arabic">
<li><p>Filter sequences by the host of interest (1 conformer vs 1 regulator?)</p>
<p>a. Should this be dropped in favour of anatomy type?</p>
</li>
<li><p>Filter sequences by anatomy type (Maybe focus on skeleton since it has more phylogenetic influences?)</p>
<p>a. Or should this be dropped instead?</p>
<blockquote>
<div><p>Or a matrix of conformer/regulator X skeleton/mucus X plasmid phylogenetic tree/regular tree OK?</p>
</div></blockquote>
<blockquote>
<div><p>Is that too many parameters?</p>
</div></blockquote>
</li>
<li><p>Per sequencing run, align bacterial genome using reference genome.</p></li>
<li><p>Group sequences by conformer/regulator (might be dropped?)</p></li>
<li><p>Group sequences by sample host.</p></li>
</ol>
<section id="for-plasmid">
<h3>For plasmid:<a class="headerlink" href="#for-plasmid" title="Permalink to this heading">#</a></h3>
<ol class="arabic simple">
<li><p>Identify plasmids.</p></li>
<li><p>Run plasmid’s phylogenetic tree within microbiome per host.</p></li>
</ol>
</section>
<section id="for-regular-tree">
<h3>For regular tree:<a class="headerlink" href="#for-regular-tree" title="Permalink to this heading">#</a></h3>
<ol class="arabic simple">
<li><p>Run phylogenetic tree within microbiome by species.</p></li>
</ol>
</section>
<section id="comparison">
<h3>Comparison:<a class="headerlink" href="#comparison" title="Permalink to this heading">#</a></h3>
<ol class="arabic simple">
<li><p>Aggregate tree based on conformer / regulator (might be dropped?)</p></li>
<li><p>Compare plasmid conformer vs regular conformer and same for regulator (?)</p></li>
</ol>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./src"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  <!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="exploratory/van_Oppen_2019.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Coral microbiome dynamics, functions and design in a changing world</p>
      </div>
    </a>
    <a class="right-next"
       href="exploratory/genome/index.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">List of used datasets for analysis</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#id4"></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#planned-procedure">Planned procedure</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#for-plasmid">For plasmid:</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#for-regular-tree">For regular tree:</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#comparison">Comparison:</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Kaede Ito
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=ac02cc09edc035673794"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=ac02cc09edc035673794"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>