

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>References &#8212; Coral Bleaching - BIOL 417</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=ac02cc09edc035673794" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=ac02cc09edc035673794" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=ac02cc09edc035673794" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=ac02cc09edc035673794" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/drawio.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=ac02cc09edc035673794" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=ac02cc09edc035673794" />
  <script src="../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=ac02cc09edc035673794"></script>

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'src/references';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="Download genomes" href="4_full_analysis/download_genome.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../readme.html">
  
  
  
  
  
    <p class="title logo__title">Coral Bleaching - BIOL 417</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../readme.html">
                    Welcome to BIOL 417 101 Capstone!
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Exploring Literature</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="1_exploratory/index.html">Proposal ideas</a></li>
<li class="toctree-l1"><a class="reference internal" href="1_exploratory/datasets.html">List of used datasets for analysis</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Proposal</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="2_proposal/index.html">Proposal</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_proposal/literature.html">Literature Review</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Pilot analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="3_pilot_1_2/index.html">Pilot 1 &amp; 2</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="3_pilot_3/download.html">Pilot 3</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="3_pilot_3/tree_finding.html">Tree finding</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Full analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="4_full_analysis/download.html">Final analysis download</a></li>
<li class="toctree-l1"><a class="reference internal" href="4_full_analysis/download_genome.html">Download genomes</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">End</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">References</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/Mapleia/BIOL417_Coral" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/src/references.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>References</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="references">
<h1>References<a class="headerlink" href="#references" title="Permalink to this heading">#</a></h1>
<div class="docutils container" id="id1">
<dl class="citation">
<dt class="label" id="id26"><span class="brackets">httpsstatsstackexchangecomusers805glenb</span></dt>
<dd><p>Glen_b (https://stats.stackexchange.com/users/805/glen-b). When do you use aic vs. bic. Cross Validated. URL:https://stats.stackexchange.com/q/57149 (version: 2013-04-25). URL: <a class="reference external" href="https://stats.stackexchange.com/q/57149">https://stats.stackexchange.com/q/57149</a>, <a class="reference external" href="https://arxiv.org/abs/https://stats.stackexchange.com/q/57149">arXiv:https://stats.stackexchange.com/q/57149</a>.</p>
</dd>
<dt class="label" id="id47"><span class="brackets">BNAH+13</span></dt>
<dd><p>Till Bayer, Matthew J. Neave, Areej Alsheikh-Hussain, Manuel Aranda, Lauren K. Yum, Tracy Mincer, Konrad Hughen, Amy Apprill, and Christian R. Voolstra. The microbiome of the red sea coral stylophora pistillata is dominated by tissue-associated endozoicomonas bacteria. <em>Applied and Environmental Microbiology</em>, 79(15):4759–4762, aug 2013. <a class="reference external" href="https://doi.org/10.1128/aem.00695-13">doi:10.1128/aem.00695-13</a>.</p>
</dd>
<dt class="label" id="id63"><span class="brackets">Bli23</span></dt>
<dd><p>Kai Blin. Ncbi-genome-download. July 2023. URL: <a class="reference external" href="https://doi.org/10.5281/zenodo.8192486">https://doi.org/10.5281/zenodo.8192486</a>, <a class="reference external" href="https://doi.org/10.5281/zenodo.8192486">doi:10.5281/zenodo.8192486</a>.</p>
</dd>
<dt class="label" id="id54"><span class="brackets">BM05</span></dt>
<dd><p>David G. Bourne and Colin B. Munn. Diversity of bacteria associated with the coral pocillopora damicornis from the great barrier reef. <em>Environmental Microbiology</em>, 7(8):1162–1174, apr 2005. <a class="reference external" href="https://doi.org/10.1111/j.1462-2920.2005.00793.x">doi:10.1111/j.1462-2920.2005.00793.x</a>.</p>
</dd>
<dt class="label" id="id36"><span class="brackets">CAC+09</span></dt>
<dd><p>Peter J. A. Cock, Tiago Antao, Jeffrey T. Chang, Brad A. Chapman, Cymon J. Cox, Andrew Dalke, Iddo Friedberg, Thomas Hamelryck, Frank Kauff, Bartek Wilczynski, and Michiel J. L. de Hoon. Biopython: freely available python tools for computational molecular biology and bioinformatics. <em>Bioinformatics</em>, 25(11):1422–1423, mar 2009. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btp163">doi:10.1093/bioinformatics/btp163</a>.</p>
</dd>
<dt class="label" id="id33"><span class="brackets">DPK+19</span></dt>
<dd><p>Diego Darriba, David Posada, Alexey M Kozlov, Alexandros Stamatakis, Benoit Morel, and Tomas Flouri. ModelTest-NG: A New and Scalable Tool for the Selection of DNA and Protein Evolutionary Models. <em>Molecular Biology and Evolution</em>, 37(1):291–294, 08 2019. URL: <a class="reference external" href="https://doi.org/10.1093/molbev/msz189">https://doi.org/10.1093/molbev/msz189</a>, <a class="reference external" href="https://arxiv.org/abs/https://academic.oup.com/mbe/article-pdf/37/1/291/32085561/msz189.pdf">arXiv:https://academic.oup.com/mbe/article-pdf/37/1/291/32085561/msz189.pdf</a>, <a class="reference external" href="https://doi.org/10.1093/molbev/msz189">doi:10.1093/molbev/msz189</a>.</p>
</dd>
<dt class="label" id="id30"><span class="brackets">Edg22</span></dt>
<dd><p>Robert C. Edgar. High-accuracy alignment ensembles enable unbiased assessments of sequence homology and phylogeny. <em>bioRxiv</em>, 2022. URL: <a class="reference external" href="https://www.biorxiv.org/content/early/2022/08/01/2021.06.20.449169">https://www.biorxiv.org/content/early/2022/08/01/2021.06.20.449169</a>, <a class="reference external" href="https://arxiv.org/abs/https://www.biorxiv.org/content/early/2022/08/01/2021.06.20.449169.full.pdf">arXiv:https://www.biorxiv.org/content/early/2022/08/01/2021.06.20.449169.full.pdf</a>, <a class="reference external" href="https://doi.org/10.1101/2021.06.20.449169">doi:10.1101/2021.06.20.449169</a>.</p>
</dd>
<dt class="label" id="id32"><span class="brackets">EKAS21</span></dt>
<dd><p>Daniel Edler, Johannes Klein, Alexandre Antonelli, and Daniele Silvestro. Raxmlgui 2.0: a graphical interface and toolkit for phylogenetic analyses using raxml. <em>Methods in Ecology and Evolution</em>, 12(2):373–377, 2021. URL: <a class="reference external" href="https://besjournals.onlinelibrary.wiley.com/doi/abs/10.1111/2041-210X.13512">https://besjournals.onlinelibrary.wiley.com/doi/abs/10.1111/2041-210X.13512</a>, <a class="reference external" href="https://arxiv.org/abs/https://besjournals.onlinelibrary.wiley.com/doi/pdf/10.1111/2041-210X.13512">arXiv:https://besjournals.onlinelibrary.wiley.com/doi/pdf/10.1111/2041-210X.13512</a>, <a class="reference external" href="https://doi.org/https://doi.org/10.1111/2041-210X.13512">doi:https://doi.org/10.1111/2041-210X.13512</a>.</p>
</dd>
<dt class="label" id="id38"><span class="brackets">Fel85</span></dt>
<dd><p>Joseph Felsenstein. Confidence limits on phylogenies: an approach using the bootstrap. <em>Evolution</em>, 39(4):783, jul 1985. <a class="reference external" href="https://doi.org/10.2307/2408678">doi:10.2307/2408678</a>.</p>
</dd>
<dt class="label" id="id31"><span class="brackets">H+99</span></dt>
<dd><p>Tom A Hall and others. Bioedit: a user-friendly biological sequence alignment editor and analysis program for windows 95/98/nt. In <em>Nucleic acids symposium series</em>, volume 41, 95–98. Oxford, 1999.</p>
</dd>
<dt class="label" id="id62"><span class="brackets">INM+22</span></dt>
<dd><p>Keigo Ide, Yohei Nishikawa, Toru Maruyama, Yuko Tsukada, Masato Kogawa, Hiroki Takeda, Haruka Ito, Ryota Wagatsuma, Rimi Miyaoka, Yoshikatsu Nakano, Koji Kinjo, Michihiro Ito, Masahito Hosokawa, Kei Yura, Shoichiro Suda, and Haruko Takeyama. Targeted single-cell genomics reveals novel host adaptation strategies of the symbiotic bacteria endozoicomonas in acropora tenuis coral. <em>Microbiome</em>, 10(1):220, Dec 2022. URL: <a class="reference external" href="https://doi.org/10.1186/s40168-022-01395-9">https://doi.org/10.1186/s40168-022-01395-9</a>, <a class="reference external" href="https://doi.org/10.1186/s40168-022-01395-9">doi:10.1186/s40168-022-01395-9</a>.</p>
</dd>
<dt class="label" id="id58"><span class="brackets">JKJ10</span></dt>
<dd><p>Yochan Joung, Haneul Kim, and Kiseong Joh. Gamma proteobacterium hmd3022 16s ribosomal rna gene, partial sequence. <em>Bioscience and Biotechnology, Hankuk University of Foreign Studies, 89, Wangsan, Mohyun, Cheoin, Yongin, Gyeonggi 449-791, Korea</em>, 2010.</p>
</dd>
<dt class="label" id="id35"><span class="brackets">KDF+19</span></dt>
<dd><p>Alexey M Kozlov, Diego Darriba, Tomáš Flouri, Benoit Morel, and Alexandros Stamatakis. RAxML-NG: a fast, scalable and user-friendly tool for maximum likelihood phylogenetic inference. <em>Bioinformatics</em>, 35(21):4453–4455, 05 2019. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btz305">https://doi.org/10.1093/bioinformatics/btz305</a>, <a class="reference external" href="https://arxiv.org/abs/https://academic.oup.com/bioinformatics/article-pdf/35/21/4453/30330793/btz305.pdf">arXiv:https://academic.oup.com/bioinformatics/article-pdf/35/21/4453/30330793/btz305.pdf</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btz305">doi:10.1093/bioinformatics/btz305</a>.</p>
</dd>
<dt class="label" id="id50"><span class="brackets">KY07</span></dt>
<dd><p>Midori Kurahashi and Akira Yokota. Endozoicomonas elysicola gen. nov., sp. nov., a γ-proteobacterium isolated from the sea slug elysia ornata. <em>Systematic and Applied Microbiology</em>, 30(3):202–206, apr 2007. <a class="reference external" href="https://doi.org/10.1016/j.syapm.2006.07.003">doi:10.1016/j.syapm.2006.07.003</a>.</p>
</dd>
<dt class="label" id="id11"><span class="brackets">LSC+18</span></dt>
<dd><p>Deborah C. A. Leite, Joana F. Salles, Emiliano N. Calderon, Jan D. van Elsas, and Raquel S. Peixoto. Specific plasmid patterns and high rates of bacterial co-occurrence within the coral holobiont. <em>Ecology and Evolution</em>, 8(3):1818–1832, jan 2018. <a class="reference external" href="https://doi.org/10.1002/ece3.3717">doi:10.1002/ece3.3717</a>.</p>
</dd>
<dt class="label" id="id57"><span class="brackets">LWW10</span></dt>
<dd><p>Heidi M. Luter, Steve Whalan, and Nicole S. Webster. Exploring the role of microorganisms in the disease-like syndrome affecting the sponge ianthella basta. <em>Applied and Environmental Microbiology</em>, 76(17):5736–5744, sep 2010. <a class="reference external" href="https://doi.org/10.1128/aem.00653-10">doi:10.1128/aem.00653-10</a>.</p>
</dd>
<dt class="label" id="id52"><span class="brackets">MGDVW+06</span></dt>
<dd><p>Manuel Martínez-García, Marta Díaz-Valdés, Gerhard Wanner, Alfonso Ramos-Esplá, and Josefa Antón. Microbial community associated with the colonial ascidian cystodytes dellechiajei. <em>Environmental Microbiology</em>, 9(2):521–534, nov 2006. <a class="reference external" href="https://doi.org/10.1111/j.1462-2920.2006.01170.x">doi:10.1111/j.1462-2920.2006.01170.x</a>.</p>
</dd>
<dt class="label" id="id49"><span class="brackets">NAK+13</span></dt>
<dd><p>Miyuki Nishijima, Kyoko Adachi, Atsuko Katsuta, Yoshikazu Shizuri, and Kazuhide Yamasato. Endozoicomonas numazuensis sp. nov., a gammaproteobacterium isolated from marine sponges, and emended description of the genus endozoicomonas kurahashi and yokota 2007. <em>International Journal of Systematic and Evolutionary Microbiology</em>, 63(Pt_2):709–714, feb 2013. <a class="reference external" href="https://doi.org/10.1099/ijs.0.042077-0">doi:10.1099/ijs.0.042077-0</a>.</p>
</dd>
<dt class="label" id="id45"><span class="brackets">pdt23</span></dt>
<dd><p>The pandas development team. Pandas-dev/pandas: pandas. November 2023. URL: <a class="reference external" href="https://doi.org/10.5281/zenodo.10107975">https://doi.org/10.5281/zenodo.10107975</a>, <a class="reference external" href="https://doi.org/10.5281/zenodo.10107975">doi:10.5281/zenodo.10107975</a>.</p>
</dd>
<dt class="label" id="id28"><span class="brackets">PHK13</span></dt>
<dd><p>Rebecca E. Pike, Brad Haltli, and Russell G. Kerr. Description of endozoicomonas euniceicola sp. nov. and endozoicomonas gorgoniicola sp. nov., bacteria isolated from the octocorals eunicea fusca and plexaura sp., and an emended description of the genus endozoicomonas. <em>International Journal of Systematic and Evolutionary Microbiology</em>, 63(Pt_11):4294–4302, nov 2013. <a class="reference external" href="https://doi.org/10.1099/ijs.0.051490-0">doi:10.1099/ijs.0.051490-0</a>.</p>
</dd>
<dt class="label" id="id44"><span class="brackets">PORadecker+22</span></dt>
<dd><p>Claudia Pogoreutz, Clinton A. Oakley, Nils Rädecker, Anny Cárdenas, Gabriela Perna, Nan Xiang, Lifeng Peng, Simon K. Davy, David K. Ngugi, and Christian R. Voolstra. Coral holobiont cues prime endozoicomonas for a symbiotic lifestyle. <em>The ISME Journal</em>, 16(8):1883–1895, Aug 2022. URL: <a class="reference external" href="https://doi.org/10.1038/s41396-022-01226-7">https://doi.org/10.1038/s41396-022-01226-7</a>, <a class="reference external" href="https://doi.org/10.1038/s41396-022-01226-7">doi:10.1038/s41396-022-01226-7</a>.</p>
</dd>
<dt class="label" id="id8"><span class="brackets">PMS+18</span></dt>
<dd><p>F. Joseph Pollock, Ryan McMinds, Styles Smith, David G. Bourne, Bette L. Willis, Mónica Medina, Rebecca Vega Thurber, and Jesse R. Zaneveld. Coral-associated bacteria demonstrate phylosymbiosis and cophylogeny. <em>Nature Communications</em>, nov 2018. <a class="reference external" href="https://doi.org/10.1038/s41467-018-07275-x">doi:10.1038/s41467-018-07275-x</a>.</p>
</dd>
<dt class="label" id="id53"><span class="brackets">RRK07</span></dt>
<dd><p>LK Ranzer, PF Restrepo, and RG Kerr. Data from: microbial community profiles of bleached and wild-type muricea elongata. ncbi popset 134140623. <em>Chemistry and Biochemistry, Florida Atlantic University, 777 Glades Rd., Boca Raton, FL 33431, USA</em>, April 2007. URL: <a class="reference external" href="https://www.ncbi.nlm.nih.gov/popset/134140623?report=genbank">https://www.ncbi.nlm.nih.gov/popset/134140623?report=genbank</a>.</p>
</dd>
<dt class="label" id="id61"><span class="brackets">SD12</span></dt>
<dd><p>Mark D. Speck and Stuart P. Donachie. Widespread oceanospirillaceae bacteria in$\less $i$\greater $porites$\less $/i$\greater $spp. <em>Journal of Marine Biology</em>, 2012:1–7, 2012. <a class="reference external" href="https://doi.org/10.1155/2012/746720">doi:10.1155/2012/746720</a>.</p>
</dd>
<dt class="label" id="id59"><span class="brackets">SDP+09</span></dt>
<dd><p>Shinichi Sunagawa, Todd Z DeSantis, Yvette M Piceno, Eoin L Brodie, Michael K DeSalvo, Christian R Voolstra, Ernesto Weil, Gary L Andersen, and Mónica Medina. Bacterial diversity and white plague disease-associated community changes in the caribbean coral montastraea faveolata. <em>The ISME Journal</em>, 3(5):512–521, jan 2009. <a class="reference external" href="https://doi.org/10.1038/ismej.2008.131">doi:10.1038/ismej.2008.131</a>.</p>
</dd>
<dt class="label" id="id48"><span class="brackets">SWM10</span></dt>
<dd><p>Shinichi Sunagawa, Cheryl M. Woodley, and Mónica Medina. Threatened corals provide underexplored microbial habitats. <em>PLoS ONE</em>, 5(3):e9554, mar 2010. <a class="reference external" href="https://doi.org/10.1371/journal.pone.0009554">doi:10.1371/journal.pone.0009554</a>.</p>
</dd>
<dt class="label" id="id37"><span class="brackets">TN93</span></dt>
<dd><p>K Tamura and M Nei. Estimation of the number of nucleotide substitutions in the control region of mitochondrial DNA in humans and chimpanzees. <em>Molecular Biology and Evolution</em>, 10(3):512–526, 05 1993. URL: <a class="reference external" href="https://doi.org/10.1093/oxfordjournals.molbev.a040023">https://doi.org/10.1093/oxfordjournals.molbev.a040023</a>, <a class="reference external" href="https://arxiv.org/abs/https://academic.oup.com/mbe/article-pdf/10/3/512/11168171/2TAMU.PDF">arXiv:https://academic.oup.com/mbe/article-pdf/10/3/512/11168171/2TAMU.PDF</a>, <a class="reference external" href="https://doi.org/10.1093/oxfordjournals.molbev.a040023">doi:10.1093/oxfordjournals.molbev.a040023</a>.</p>
</dd>
<dt class="label" id="id27"><span class="brackets">TDNK07</span></dt>
<dd><p>K. Tamura, J. Dudley, M. Nei, and S. Kumar. MEGA4: molecular evolutionary genetics analysis (MEGA) software version 4.0. <em>Molecular Biology and Evolution</em>, 24(8):1596–1599, apr 2007. <a class="reference external" href="https://doi.org/10.1093/molbev/msm092">doi:10.1093/molbev/msm092</a>.</p>
</dd>
<dt class="label" id="id29"><span class="brackets">TSK21</span></dt>
<dd><p>Koichiro Tamura, Glen Stecher, and Sudhir Kumar. MEGA11: molecular evolutionary genetics analysis version 11. <em>Molecular Biology and Evolution</em>, 38(7):3022–3027, apr 2021. <a class="reference external" href="https://doi.org/10.1093/molbev/msab120">doi:10.1093/molbev/msab120</a>.</p>
</dd>
<dt class="label" id="id51"><span class="brackets">TLS+07</span></dt>
<dd><p>Vera Thiel, Sven Leininger, Rolf Schmaljohann, Franz Brümmer, and Johannes F. Imhoff. Sponge-specific bacterial associations of the mediterranean sponge chondrilla nucula (demospongiae, tetractinomorpha). <em>Microbial Ecology</em>, 54(1):101–111, mar 2007. <a class="reference external" href="https://doi.org/10.1007/s00248-006-9177-y">doi:10.1007/s00248-006-9177-y</a>.</p>
</dd>
<dt class="label" id="id43"><span class="brackets">WHT+22</span></dt>
<dd><p>Naohisa Wada, Ming-Tsung Hsu, Kshitij Tandon, Silver Sung-Yun Hsiao, Hsing-Ju Chen, Yu-Hsiang Chen, Pei-Wen Chiang, Sheng-Ping Yu, Chih-Ying Lu, Yu-Jing Chiou, Yung-Chi Tu, Xuejiao Tian, Bi-Chang Chen, Der-Chuen Lee, Hideyuki Yamashiro, David G. Bourne, and Sen-Lin Tang. High-resolution spatial and genomic characterization of coral-associated microbial aggregates in the coral stylophora pistillata. <em>Science Advances</em>, jul 2022. <a class="reference external" href="https://doi.org/10.1126/sciadv.abo2431">doi:10.1126/sciadv.abo2431</a>.</p>
</dd>
<dt class="label" id="id60"><span class="brackets">YCA+10</span></dt>
<dd><p>Cho-Song Yang, Ming-Hui Chen, A. B. Arun, Chaolun Allen Chen, Jih-Terng Wang, and Wen-Ming Chen. Endozoicomonas montiporae sp. nov., isolated from the encrusting pore coral montipora aequituberculata. <em>International Journal of Systematic and Evolutionary Microbiology</em>, 60(5):1158–1162, may 2010. <a class="reference external" href="https://doi.org/10.1099/ijs.0.014357-0">doi:10.1099/ijs.0.014357-0</a>.</p>
</dd>
<dt class="label" id="id9"><span class="brackets">ZGB+19</span></dt>
<dd><p>Maren Ziegler, Carsten G. B. Grupstra, Marcelle M. Barreto, Martin Eaton, Jaafar BaOmar, Khalid Zubier, Abdulmohsin Al-Sofyani, Adnan J. Turki, Rupert Ormond, and Christian R. Voolstra. Coral bacterial community structure responds to environmental change in a host-specific manner. <em>Nature Communications</em>, jul 2019. <a class="reference external" href="https://doi.org/10.1038/s41467-019-10969-5">doi:10.1038/s41467-019-10969-5</a>.</p>
</dd>
<dt class="label" id="id56"><span class="brackets">ZPD+09</span></dt>
<dd><p>Frank U. Zielinski, Annelie Pernthaler, Sébastien Duperron, Luciana Raggi, Olav Giere, Christian Borowski, and Nicole Dubilier. Widespread occurrence of an intranuclear bacterial parasite in vent and seep bathymodiolin mussels. <em>Environmental Microbiology</em>, 11(5):1150–1167, apr 2009. <a class="reference external" href="https://doi.org/10.1111/j.1462-2920.2008.01847.x">doi:10.1111/j.1462-2920.2008.01847.x</a>.</p>
</dd>
<dt class="label" id="id46"><span class="brackets">WesMcKinney10</span></dt>
<dd><p>Wes McKinney. Data Structures for Statistical Computing in Python. In Stéfan van der Walt and Jarrod Millman, editors, <em>Proceedings of the 9th Python in Science Conference</em>, 56 – 61. 2010. <a class="reference external" href="https://doi.org/10.25080/Majora-92bf1922-00a">doi:10.25080/Majora-92bf1922-00a</a>.</p>
</dd>
</dl>
</div>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./src"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  <!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="4_full_analysis/download_genome.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Download genomes</p>
      </div>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Kaede Ito
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=ac02cc09edc035673794"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=ac02cc09edc035673794"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>